{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gbouras13/pharokka/blob/master/run_pharokka_and_phold_and_phynteny.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QGd2GEI3N-02"
      },
      "source": [
        "#Pharokka + Phold + Phynteny\n",
        "\n",
        "[pharokka](https://github.com/gbouras13/pharokka) is a rapid standardised annotation tool for bacteriophage genomes and metagenomes. You can read more about pharokka in the [documentation](https://pharokka.readthedocs.io/).\n",
        "\n",
        "[phold](https://github.com/gbouras13/phold) is a sensitive annotation tool for bacteriophage genomes and metagenomes using protein structural homology. You can read more about phold in the [documentation](https://phold.readthedocs.io/).\n",
        "\n",
        "phold uses the [ProstT5](https://github.com/mheinzinger/ProstT5) protein language model to translate protein amino acid sequences to the 3Di token alphabet used by [Foldseek](https://github.com/steineggerlab/foldseek). Foldseek is then used to search these against a database of 803k protein structures mostly predicted using [Colabfold](https://github.com/sokrypton/ColabFold).\n",
        "\n",
        "[phyntney](https://github.com/susiegriggo/Phynteny) uses a long-short term memory model trained on phage synteny (the conserved gene order across phages) to assign hypothetical phage proteins to a PHROG category.\n",
        "\n",
        "**NOTE: Phynteny will only work if your phage has fewer than 120 predicted genes**\n",
        "\n",
        "**If this is the case for your phage(s), you should just skip running Phynteny (Cells 5+6)**\n",
        "\n",
        "The tools are best run sequentially, as Pharokka conducts extra annotation steps like tRNA, tmRNA, CRISPR and INPHARED searches that Phold lacks (for now at least). Pharokka will also (rarely) annotate CDS that Phold can miss. Phynteny can then help annotate remaining hypothetical proteins with a PHROG category.\n",
        "\n",
        "* **Before you start, please make sure you change the runtime to T4 GPU (or any other kind of GPU if you have $$$), otherwise Phold won't be installed properly**\n",
        "* To do this, go to the top toolbar, then to Runtime -> Change runtime type -> Hardware accelerator\n",
        "\n",
        "* To run the cells, press the play button on the left side\n",
        "* Cells 1 and 2 install pharokka and phold and download the databases/models.\n",
        "* Once they have been run, you can re-run Cell 3 (to run Pharokka), Cell 4 (to run Phold) and Cell 5+6 (to install and run Phynteny) as many times as you would like\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "Ii39RG8eOZUx",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e6c390ad-965b-43f6-9dae-8e8ea0be75da"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "python version 3.10\n",
            "installing python\n",
            "installing pharokka and phold\n",
            "Channels:\n",
            " - conda-forge\n",
            " - bioconda\n",
            " - defaults\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - pharokka==1.7.3\n",
            "    - phold==0.2.0\n",
            "    - pip\n",
            "    - python=3.10\n",
            "    - pytorch[build=cuda*]\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge\n",
            "    _openmp_mutex-4.5          |       2_kmp_llvm           6 KB  conda-forge\n",
            "    about-time-4.2.1           |     pyhd8ed1ab_0          16 KB  conda-forge\n",
            "    aiohttp-3.9.5              |  py310h2372a71_0         682 KB  conda-forge\n",
            "    aiosignal-1.3.1            |     pyhd8ed1ab_0          12 KB  conda-forge\n",
            "    alive-progress-3.1.5       |     pyhd8ed1ab_0          65 KB  conda-forge\n",
            "    alsa-lib-1.2.6.1           |       h7f98852_0         578 KB  conda-forge\n",
            "    anaconda-anon-usage-0.4.4  |py310hfc0e8ea_100          25 KB\n",
            "    aragorn-1.2.41             |       h031d066_3         143 KB  bioconda\n",
            "    aria2-1.36.0               |       h1e4e653_3         1.8 MB  conda-forge\n",
            "    async-timeout-4.0.3        |     pyhd8ed1ab_0          11 KB  conda-forge\n",
            "    attrs-23.2.0               |     pyh71513ae_0          53 KB  conda-forge\n",
            "    aws-c-auth-0.7.20          |       h5f1c8d9_0         103 KB  conda-forge\n",
            "    aws-c-cal-0.6.12           |       h2ba76a8_0          45 KB  conda-forge\n",
            "    aws-c-common-0.9.17        |       h4ab18f5_0         222 KB  conda-forge\n",
            "    aws-c-compression-0.2.18   |       h36a0aea_4          19 KB  conda-forge\n",
            "    aws-c-event-stream-0.4.2   |      h161de36_10          53 KB  conda-forge\n",
            "    aws-c-http-0.8.1           |      h63f54a0_13         191 KB  conda-forge\n",
            "    aws-c-io-0.14.8            |       h96d4d28_0         154 KB  conda-forge\n",
            "    aws-c-mqtt-0.10.4          |       hcc7299c_2         160 KB  conda-forge\n",
            "    aws-c-s3-0.5.9             |       h10bd90f_0         107 KB  conda-forge\n",
            "    aws-c-sdkutils-0.1.16      |       h36a0aea_0          54 KB  conda-forge\n",
            "    aws-checksums-0.1.18       |       h36a0aea_4          49 KB  conda-forge\n",
            "    aws-crt-cpp-0.26.8         |      h4f3a3cc_11         332 KB  conda-forge\n",
            "    aws-sdk-cpp-1.11.329       |       h51dfee4_0         3.5 MB  conda-forge\n",
            "    backports-1.0              |     pyhd8ed1ab_3           6 KB  conda-forge\n",
            "    backports.tempfile-1.0     |             py_0          10 KB  conda-forge\n",
            "    backports.weakref-1.0.post1|  pyhd8ed1ab_1003          11 KB  conda-forge\n",
            "    bcbio-gff-0.7.1            |     pyh7cba7a3_1          22 KB  bioconda\n",
            "    biopython-1.81             |  py310h2372a71_1         2.7 MB  conda-forge\n",
            "    black-24.4.2               |  py310hff52083_0         296 KB  conda-forge\n",
            "    blast-2.15.0               | pl5321h6f7f691_1       146.0 MB  bioconda\n",
            "    boltons-24.0.0             |     pyhd8ed1ab_0         291 KB  conda-forge\n",
            "    brotli-1.1.0               |       hd590300_1          19 KB  conda-forge\n",
            "    brotli-bin-1.1.0           |       hd590300_1          19 KB  conda-forge\n",
            "    brotli-python-1.1.0        |  py310hc6cd4ac_1         341 KB  conda-forge\n",
            "    bx-python-0.12.0           |  py310ha025fb0_0         1.1 MB  bioconda\n",
            "    c-ares-1.32.2              |       h4bc722e_0         176 KB  conda-forge\n",
            "    ca-certificates-2024.7.4   |       hbcca054_0         151 KB  conda-forge\n",
            "    cached-property-1.5.2      |       hd8ed1ab_1           4 KB  conda-forge\n",
            "    cached_property-1.5.2      |     pyha770c72_1          11 KB  conda-forge\n",
            "    cairo-1.16.0               |       hb05425b_5         1.2 MB\n",
            "    capnproto-1.0.2            |       h2b92303_1         3.8 MB  conda-forge\n",
            "    certifi-2024.7.4           |     pyhd8ed1ab_0         156 KB  conda-forge\n",
            "    cffi-1.16.0                |  py310h2fee648_0         236 KB  conda-forge\n",
            "    click-8.1.7                |unix_pyh707e725_0          82 KB  conda-forge\n",
            "    colorama-0.4.6             |     pyhd8ed1ab_0          25 KB  conda-forge\n",
            "    conda-24.5.0               |  py310hff52083_0         939 KB  conda-forge\n",
            "    conda-content-trust-0.2.0  |     pyhd8ed1ab_0          57 KB  conda-forge\n",
            "    conda-package-handling-2.3.0|     pyh7900ff3_0         250 KB  conda-forge\n",
            "    conda-package-streaming-0.10.0|     pyhd8ed1ab_0          19 KB  conda-forge\n",
            "    contourpy-1.2.1            |  py310hd41b1e2_0         236 KB  conda-forge\n",
            "    coverage-7.6.0             |  py310h5b4e0ec_0         286 KB  conda-forge\n",
            "    cryptography-42.0.8        |  py310hb1bd9d3_0         1.9 MB  conda-forge\n",
            "    cuda-cudart-12.5.82        |       he02047a_0          22 KB  conda-forge\n",
            "    cuda-cudart_linux-64-12.5.82|       h85509e4_0         184 KB  conda-forge\n",
            "    cuda-nvrtc-12.5.82         |       he02047a_0        18.1 MB  conda-forge\n",
            "    cuda-nvtx-12.5.82          |       he02047a_0          31 KB  conda-forge\n",
            "    cuda-version-12.5          |       hd4f0392_3          21 KB  conda-forge\n",
            "    cudnn-8.9.7.29             |       h092f7fd_3       446.6 MB  conda-forge\n",
            "    curl-8.7.1                 |       hdbd6064_0          86 KB\n",
            "    cycler-0.12.1              |     pyhd8ed1ab_0          13 KB  conda-forge\n",
            "    datasets-2.20.0            |     pyhd8ed1ab_0         354 KB  conda-forge\n",
            "    dill-0.3.8                 |     pyhd8ed1ab_0          86 KB  conda-forge\n",
            "    distro-1.9.0               |     pyhd8ed1ab_0          41 KB  conda-forge\n",
            "    dnaapler-0.7.0             |     pyhdfd78af_0        10.2 MB  bioconda\n",
            "    entrez-direct-22.1         |       he881be0_0        14.1 MB  bioconda\n",
            "    exceptiongroup-1.2.2       |     pyhd8ed1ab_0          20 KB  conda-forge\n",
            "    fastpath-1.9               |  py310h2372a71_3          59 KB  conda-forge\n",
            "    filelock-3.15.4            |     pyhd8ed1ab_0          17 KB  conda-forge\n",
            "    foldseek-9.427df8a         | pl5321hb365157_1         7.6 MB  bioconda\n",
            "    font-ttf-dejavu-sans-mono-2.37|       hab24e00_0         388 KB  conda-forge\n",
            "    font-ttf-inconsolata-3.000 |       h77eed37_0          94 KB  conda-forge\n",
            "    font-ttf-source-code-pro-2.038|       h77eed37_0         684 KB  conda-forge\n",
            "    font-ttf-ubuntu-0.83       |       h77eed37_2         1.5 MB  conda-forge\n",
            "    fontconfig-2.14.2          |       h14ed4e7_0         266 KB  conda-forge\n",
            "    fonts-conda-ecosystem-1    |                0           4 KB  conda-forge\n",
            "    fonts-conda-forge-1        |                0           4 KB  conda-forge\n",
            "    fonttools-4.53.1           |  py310h5b4e0ec_0         2.3 MB  conda-forge\n",
            "    freetype-2.12.1            |       h267a509_2         620 KB  conda-forge\n",
            "    frozendict-2.4.4           |  py310hc51659f_0          48 KB  conda-forge\n",
            "    frozenlist-1.4.1           |  py310h2372a71_0          58 KB  conda-forge\n",
            "    fsspec-2024.5.0            |     pyhff2d567_0         211 KB  conda-forge\n",
            "    gawk-5.3.0                 |       ha916aea_0         1.1 MB  conda-forge\n",
            "    gettext-0.22.5             |       h59595ed_2         464 KB  conda-forge\n",
            "    gettext-tools-0.22.5       |       h59595ed_2         2.6 MB  conda-forge\n",
            "    gflags-2.2.2               |    he1b5a44_1004         114 KB  conda-forge\n",
            "    giflib-5.2.2               |       hd590300_0          75 KB  conda-forge\n",
            "    glib-2.78.4                |       hfc55251_0         478 KB  conda-forge\n",
            "    glib-tools-2.78.4          |       hfc55251_0         109 KB  conda-forge\n",
            "    glog-0.7.1                 |       hbabe93e_0         140 KB  conda-forge\n",
            "    gmp-6.3.0                  |       hac33072_2         449 KB  conda-forge\n",
            "    gmpy2-2.1.5                |  py310hc7909c9_1         201 KB  conda-forge\n",
            "    grapheme-0.6.0             |     pyhd8ed1ab_0         141 KB  conda-forge\n",
            "    graphite2-1.3.14           |       h295c915_1          97 KB\n",
            "    gsl-2.7                    |       he838d99_0         3.2 MB  conda-forge\n",
            "    h2-4.1.0                   |     pyhd8ed1ab_0          46 KB  conda-forge\n",
            "    h5py-3.11.0                |nompi_py310hf054cd7_102         1.1 MB  conda-forge\n",
            "    harfbuzz-4.3.0             |       hf52aaf7_2         1.3 MB\n",
            "    hdf5-1.14.3                |nompi_h4f84152_101         3.7 MB  conda-forge\n",
            "    hpack-4.0.0                |     pyh9f0ad1d_0          25 KB  conda-forge\n",
            "    huggingface_hub-0.23.4     |     pyhd8ed1ab_0         245 KB  conda-forge\n",
            "    hyperframe-6.0.1           |     pyhd8ed1ab_0          14 KB  conda-forge\n",
            "    idna-3.7                   |     pyhd8ed1ab_0          51 KB  conda-forge\n",
            "    infernal-1.1.5             | pl5321h031d066_2         3.7 MB  bioconda\n",
            "    iniconfig-2.0.0            |     pyhd8ed1ab_0          11 KB  conda-forge\n",
            "    isort-5.13.2               |     pyhd8ed1ab_0          72 KB  conda-forge\n",
            "    jinja2-3.1.4               |     pyhd8ed1ab_0         109 KB  conda-forge\n",
            "    jpeg-9e                    |       h0b41bf4_3         235 KB  conda-forge\n",
            "    jsonpatch-1.33             |     pyhd8ed1ab_0          17 KB  conda-forge\n",
            "    kiwisolver-1.4.5           |  py310hd41b1e2_1          71 KB  conda-forge\n",
            "    lcms2-2.15                 |       hfd0df8a_0         235 KB  conda-forge\n",
            "    lerc-4.0.0                 |       h27087fc_0         275 KB  conda-forge\n",
            "    libabseil-20240116.2       | cxx17_he02047a_1         1.2 MB  conda-forge\n",
            "    libaec-1.1.3               |       h59595ed_0          35 KB  conda-forge\n",
            "    libarrow-16.1.0            |   h06228bf_5_cpu         7.9 MB  conda-forge\n",
            "    libarrow-acero-16.1.0      |   hac33072_5_cpu         586 KB  conda-forge\n",
            "    libarrow-dataset-16.1.0    |   hac33072_5_cpu         567 KB  conda-forge\n",
            "    libarrow-substrait-16.1.0  |   h7e0c224_5_cpu         540 KB  conda-forge\n",
            "    libasprintf-0.22.5         |       h661eb56_2          42 KB  conda-forge\n",
            "    libasprintf-devel-0.22.5   |       h661eb56_2          33 KB  conda-forge\n",
            "    libblas-3.9.0              |22_linux64_openblas          14 KB  conda-forge\n",
            "    libbrotlicommon-1.1.0      |       hd590300_1          68 KB  conda-forge\n",
            "    libbrotlidec-1.1.0         |       hd590300_1          32 KB  conda-forge\n",
            "    libbrotlienc-1.1.0         |       hd590300_1         276 KB  conda-forge\n",
            "    libcblas-3.9.0             |22_linux64_openblas          14 KB  conda-forge\n",
            "    libcrc32c-1.1.2            |       h9c3ff4c_0          20 KB  conda-forge\n",
            "    libcublas-12.5.3.2         |       he02047a_0       240.0 MB  conda-forge\n",
            "    libcufft-11.2.3.61         |       he02047a_0       148.6 MB  conda-forge\n",
            "    libcups-2.3.3              |       h36d4200_3         4.3 MB  conda-forge\n",
            "    libcurand-10.3.6.82        |       he02047a_0        39.8 MB  conda-forge\n",
            "    libcusolver-11.6.3.83      |       he02047a_0        79.3 MB  conda-forge\n",
            "    libcusparse-12.5.1.3       |       he02047a_0       119.1 MB  conda-forge\n",
            "    libdeflate-1.17            |       h0b41bf4_0          63 KB  conda-forge\n",
            "    libevent-2.1.12            |       hf998b51_1         417 KB  conda-forge\n",
            "    libgcc-ng-14.1.0           |       h77fa898_0         822 KB  conda-forge\n",
            "    libgettextpo-0.22.5        |       h59595ed_2         167 KB  conda-forge\n",
            "    libgettextpo-devel-0.22.5  |       h59595ed_2          36 KB  conda-forge\n",
            "    libgfortran-ng-14.1.0      |       h69a702a_0          49 KB  conda-forge\n",
            "    libgfortran5-14.1.0        |       hc5f4f2c_0         1.4 MB  conda-forge\n",
            "    libglib-2.78.4             |       h783c2da_0         2.6 MB  conda-forge\n",
            "    libgomp-14.1.0             |       h77fa898_0         446 KB  conda-forge\n",
            "    libgoogle-cloud-2.24.0     |       h2736e30_0         1.2 MB  conda-forge\n",
            "    libgoogle-cloud-storage-2.24.0|       h3d9a0c8_0         735 KB  conda-forge\n",
            "    libgrpc-1.62.2             |       h15f2491_0         7.0 MB  conda-forge\n",
            "    libhwloc-2.9.1             |       hd6dc26d_0         2.5 MB  conda-forge\n",
            "    libiconv-1.17              |       hd590300_2         689 KB  conda-forge\n",
            "    libidn2-2.3.7              |       hd590300_0         124 KB  conda-forge\n",
            "    liblapack-3.9.0            |22_linux64_openblas          14 KB  conda-forge\n",
            "    libmagma-2.7.2             |       h173bb3b_2       229.9 MB  conda-forge\n",
            "    libmagma_sparse-2.7.2      |       h173bb3b_3         6.4 MB  conda-forge\n",
            "    libmambapy-1.5.8           |  py310h2dafd23_2         327 KB\n",
            "    libnsl-2.0.1               |       hd590300_0          33 KB  conda-forge\n",
            "    libnvjitlink-12.5.82       |       he02047a_0        15.9 MB  conda-forge\n",
            "    libopenblas-0.3.27         |pthreads_hac2b453_1         5.3 MB  conda-forge\n",
            "    libparquet-16.1.0          |   h6a7eafb_5_cpu         1.1 MB  conda-forge\n",
            "    libpng-1.6.43              |       h2797004_0         281 KB  conda-forge\n",
            "    libprotobuf-4.25.3         |       h08a7969_0         2.7 MB  conda-forge\n",
            "    libre2-11-2023.09.01       |       h5a48ba9_2         227 KB  conda-forge\n",
            "    libsentencepiece-0.2.0     |       he81a138_2         801 KB  conda-forge\n",
            "    libsqlite-3.46.0           |       hde9e2c9_0         845 KB  conda-forge\n",
            "    libstdcxx-ng-14.1.0        |       hc0a3c3a_0         3.7 MB  conda-forge\n",
            "    libthrift-0.19.0           |       hb90f79a_1         400 KB  conda-forge\n",
            "    libtiff-4.5.0              |       h6adf6a1_2         397 KB  conda-forge\n",
            "    libtorch-2.3.1             |cuda120_h2b0da52_300       481.8 MB  conda-forge\n",
            "    libunistring-0.9.10        |       h7f98852_0         1.4 MB  conda-forge\n",
            "    libutf8proc-2.8.0          |       h166bdaf_0          99 KB  conda-forge\n",
            "    libuuid-2.38.1             |       h0b41bf4_0          33 KB  conda-forge\n",
            "    libuv-1.48.0               |       hd590300_0         879 KB  conda-forge\n",
            "    libwebp-base-1.4.0         |       hd590300_0         429 KB  conda-forge\n",
            "    libxcb-1.16                |       hd590300_0         386 KB  conda-forge\n",
            "    libxcrypt-4.4.36           |       hd590300_1          98 KB  conda-forge\n",
            "    libzlib-1.2.13             |       h4ab18f5_6          60 KB  conda-forge\n",
            "    llvm-openmp-18.1.7         |       ha31de31_0        55.9 MB  conda-forge\n",
            "    loguru-0.7.2               |  py310hff52083_1          96 KB  conda-forge\n",
            "    markupsafe-2.1.5           |  py310h2372a71_0          24 KB  conda-forge\n",
            "    mash-2.3                   |       hc74b729_7         588 KB  bioconda\n",
            "    matplotlib-base-3.9.1      |  py310h0b1de36_0         6.7 MB  conda-forge\n",
            "    menuinst-2.1.1             |  py310hff52083_0         134 KB  conda-forge\n",
            "    minced-0.4.2               |       hdfd78af_1          47 KB  bioconda\n",
            "    mkl-2023.2.0               |   h84fe81f_50496       156.8 MB  conda-forge\n",
            "    mmseqs2-13.45111           | pl5321hf1761c0_2         2.1 MB  bioconda\n",
            "    mpc-1.3.1                  |       hfe3b2da_0         114 KB  conda-forge\n",
            "    mpfr-4.2.1                 |       h9458935_1         628 KB  conda-forge\n",
            "    mpmath-1.3.0               |     pyhd8ed1ab_0         428 KB  conda-forge\n",
            "    multidict-6.0.5            |  py310h2372a71_0          56 KB  conda-forge\n",
            "    multiprocess-0.70.16       |  py310h2372a71_0         238 KB  conda-forge\n",
            "    munkres-1.1.4              |     pyh9f0ad1d_0          12 KB  conda-forge\n",
            "    mypy_extensions-1.0.0      |     pyha770c72_0          10 KB  conda-forge\n",
            "    ncbi-vdb-3.1.1             |       h4ac6f70_1        10.7 MB  bioconda\n",
            "    nccl-2.22.3.1              |       hbc370b7_0       107.3 MB  conda-forge\n",
            "    ncurses-6.5                |       h59595ed_0         867 KB  conda-forge\n",
            "    networkx-3.3               |     pyhd8ed1ab_1         1.1 MB  conda-forge\n",
            "    numpy-2.0.0                |  py310h515e003_0         7.4 MB  conda-forge\n",
            "    openjdk-11.0.15            |       hc6918da_0       174.0 MB  conda-forge\n",
            "    openjpeg-2.5.0             |       hfec8fc6_2         344 KB  conda-forge\n",
            "    openssl-3.3.1              |       h4bc722e_2         2.8 MB  conda-forge\n",
            "    orc-2.0.1                  |       h17fec99_1        1006 KB  conda-forge\n",
            "    packaging-24.1             |     pyhd8ed1ab_0          49 KB  conda-forge\n",
            "    pandas-2.2.2               |  py310hf9f9076_1        12.4 MB  conda-forge\n",
            "    pathspec-0.12.1            |     pyhd8ed1ab_0          40 KB  conda-forge\n",
            "    patsy-0.5.6                |     pyhd8ed1ab_0         183 KB  conda-forge\n",
            "    pcre-8.45                  |       h9c3ff4c_0         253 KB  conda-forge\n",
            "    perl-5.32.1                | 7_hd590300_perl5        12.7 MB  conda-forge\n",
            "    perl-archive-tar-2.40      | pl5321hdfd78af_0          33 KB  bioconda\n",
            "    perl-carp-1.50             | pl5321hd8ed1ab_0          22 KB  conda-forge\n",
            "    perl-common-sense-3.75     | pl5321hd8ed1ab_0          20 KB  conda-forge\n",
            "    perl-compress-raw-bzip2-2.201| pl5321h166bdaf_0          54 KB  conda-forge\n",
            "    perl-compress-raw-zlib-2.202| pl5321h166bdaf_0          83 KB  conda-forge\n",
            "    perl-encode-3.21           | pl5321hd590300_0         1.7 MB  conda-forge\n",
            "    perl-exporter-5.74         | pl5321hd8ed1ab_0          19 KB  conda-forge\n",
            "    perl-exporter-tiny-1.002002| pl5321hd8ed1ab_0          28 KB  conda-forge\n",
            "    perl-extutils-makemaker-7.70| pl5321hd8ed1ab_0         154 KB  conda-forge\n",
            "    perl-io-compress-2.201     | pl5321hdbdd923_2          84 KB  bioconda\n",
            "    perl-io-zlib-1.14          | pl5321hdfd78af_0          12 KB  bioconda\n",
            "    perl-json-4.10             | pl5321hdfd78af_0          56 KB  bioconda\n",
            "    perl-json-xs-2.34          | pl5321h4ac6f70_6          66 KB  bioconda\n",
            "    perl-list-moreutils-0.430  | pl5321hdfd78af_0          32 KB  bioconda\n",
            "    perl-list-moreutils-xs-0.430| pl5321h031d066_2          50 KB  bioconda\n",
            "    perl-parent-0.241          | pl5321hd8ed1ab_0          13 KB  conda-forge\n",
            "    perl-pathtools-3.75        | pl5321h166bdaf_0          49 KB  conda-forge\n",
            "    perl-scalar-list-utils-1.63| pl5321h166bdaf_0          50 KB  conda-forge\n",
            "    perl-storable-3.15         | pl5321h166bdaf_0          70 KB  conda-forge\n",
            "    perl-types-serialiser-1.01 | pl5321hdfd78af_0          13 KB  bioconda\n",
            "    phanotate-1.5.1            |       h0dbaff4_1          57 KB  bioconda\n",
            "    pharokka-1.7.3             |     pyhdfd78af_0          74 KB  bioconda\n",
            "    phold-0.2.0                |     pyhdfd78af_0        12.4 MB  bioconda\n",
            "    pillow-10.4.0              |  py310h5eee18b_0         804 KB\n",
            "    pip-24.0                   |     pyhd8ed1ab_0         1.3 MB  conda-forge\n",
            "    pixman-0.43.2              |       h59595ed_0         378 KB  conda-forge\n",
            "    platformdirs-4.2.2         |     pyhd8ed1ab_0          20 KB  conda-forge\n",
            "    pluggy-1.5.0               |     pyhd8ed1ab_0          23 KB  conda-forge\n",
            "    psutil-6.0.0               |  py310hc51659f_0         363 KB  conda-forge\n",
            "    pthread-stubs-0.4          |    h36c2ea0_1001           5 KB  conda-forge\n",
            "    pyarrow-16.1.0             |  py310hb7f781d_2          28 KB  conda-forge\n",
            "    pyarrow-core-16.1.0        |py310hcccf80d_2_cpu         4.2 MB  conda-forge\n",
            "    pyarrow-hotfix-0.6         |     pyhd8ed1ab_0          13 KB  conda-forge\n",
            "    pybind11-abi-4             |       hd8ed1ab_3          10 KB  conda-forge\n",
            "    pycirclize-1.6.0           |     pyhd8ed1ab_0          62 KB  conda-forge\n",
            "    pycosat-0.6.6              |  py310h2372a71_0          85 KB  conda-forge\n",
            "    pyhmmer-0.10.14            |  py310h7c593f9_0         4.6 MB  bioconda\n",
            "    pyparsing-3.1.2            |     pyhd8ed1ab_0          87 KB  conda-forge\n",
            "    pyrodigal-3.5.0            |  py310h7c593f9_0         2.6 MB  bioconda\n",
            "    pyrodigal-gv-0.3.1         |     pyh7cba7a3_0         575 KB  bioconda\n",
            "    pysocks-1.7.1              |     pyha2e5f31_6          19 KB  conda-forge\n",
            "    pytest-8.2.2               |     pyhd8ed1ab_0         251 KB  conda-forge\n",
            "    pytest-cov-5.0.0           |     pyhd8ed1ab_0          25 KB  conda-forge\n",
            "    python-3.10.14             |hd12c33a_0_cpython        24.3 MB  conda-forge\n",
            "    python-dateutil-2.9.0      |     pyhd8ed1ab_0         218 KB  conda-forge\n",
            "    python-tzdata-2024.1       |     pyhd8ed1ab_0         141 KB  conda-forge\n",
            "    python-xxhash-3.4.1        |  py310h2372a71_0          23 KB  conda-forge\n",
            "    python_abi-3.10            |          4_cp310           6 KB  conda-forge\n",
            "    pytorch-2.3.1              |cuda120_py310h2c91c31_300        29.0 MB  conda-forge\n",
            "    pytz-2024.1                |     pyhd8ed1ab_0         184 KB  conda-forge\n",
            "    pyyaml-6.0.1               |  py310h2372a71_1         167 KB  conda-forge\n",
            "    qhull-2020.2               |       h434a139_5         540 KB  conda-forge\n",
            "    re2-2023.09.01             |       h7f4b329_2          26 KB  conda-forge\n",
            "    regex-2024.5.15            |  py310hc51659f_0         340 KB  conda-forge\n",
            "    requests-2.32.3            |     pyhd8ed1ab_0          57 KB  conda-forge\n",
            "    ruamel.yaml-0.18.6         |  py310h2372a71_0         199 KB  conda-forge\n",
            "    ruamel.yaml.clib-0.2.8     |  py310h2372a71_0         133 KB  conda-forge\n",
            "    s2n-1.4.13                 |       he19d79f_0         339 KB  conda-forge\n",
            "    safetensors-0.4.3          |  py310hcb5633a_0         966 KB  conda-forge\n",
            "    scipy-1.14.0               |  py310h93e2701_1        16.0 MB  conda-forge\n",
            "    seaborn-0.13.2             |       hd8ed1ab_2           7 KB  conda-forge\n",
            "    seaborn-base-0.13.2        |     pyhd8ed1ab_2         229 KB  conda-forge\n",
            "    sentencepiece-0.2.0        |       hff52083_2          16 KB  conda-forge\n",
            "    sentencepiece-python-0.2.0 |  py310h91407d4_2         2.5 MB  conda-forge\n",
            "    sentencepiece-spm-0.2.0    |       he81a138_2          83 KB  conda-forge\n",
            "    setuptools-70.3.0          |     pyhd8ed1ab_0         485 KB  conda-forge\n",
            "    six-1.16.0                 |     pyh6c4a22f_0          14 KB  conda-forge\n",
            "    sleef-3.6.1                |       h3400bea_1         975 KB  conda-forge\n",
            "    snappy-1.2.1               |       ha2e4443_0          41 KB  conda-forge\n",
            "    statsmodels-0.14.2         |  py310h261611a_0        10.4 MB  conda-forge\n",
            "    sympy-1.13.0               | pypyh2585a3b_103         4.3 MB  conda-forge\n",
            "    tbb-2021.9.0               |       hf52228f_0         1.5 MB  conda-forge\n",
            "    textwrap3-0.9.2            |             py_0          12 KB  conda-forge\n",
            "    tokenizers-0.19.1          |  py310h320607d_0         2.6 MB  conda-forge\n",
            "    toml-0.10.2                |     pyhd8ed1ab_0          18 KB  conda-forge\n",
            "    tomli-2.0.1                |     pyhd8ed1ab_0          16 KB  conda-forge\n",
            "    tqdm-4.66.4                |     pyhd8ed1ab_0          87 KB  conda-forge\n",
            "    transformers-4.41.2        |     pyhd8ed1ab_0         3.1 MB  conda-forge\n",
            "    trnascan-se-2.0.12         | pl5321h031d066_0         3.1 MB  bioconda\n",
            "    truststore-0.8.0           |     pyhd8ed1ab_0          20 KB  conda-forge\n",
            "    typing-extensions-4.12.2   |       hd8ed1ab_0          10 KB  conda-forge\n",
            "    typing_extensions-4.12.2   |     pyha770c72_0          39 KB  conda-forge\n",
            "    unicodedata2-15.1.0        |  py310h2372a71_0         365 KB  conda-forge\n",
            "    urllib3-2.2.2              |     pyhd8ed1ab_1          93 KB  conda-forge\n",
            "    wget-1.21.4                |       hda4d442_0         752 KB  conda-forge\n",
            "    wheel-0.43.0               |     pyhd8ed1ab_1          57 KB  conda-forge\n",
            "    xorg-fixesproto-5.0        |    h7f98852_1002           9 KB  conda-forge\n",
            "    xorg-inputproto-2.3.2      |    h7f98852_1002          19 KB  conda-forge\n",
            "    xorg-kbproto-1.0.7         |    h7f98852_1002          27 KB  conda-forge\n",
            "    xorg-libx11-1.8.9          |       hb711507_1         813 KB  conda-forge\n",
            "    xorg-libxau-1.0.11         |       hd590300_0          14 KB  conda-forge\n",
            "    xorg-libxdmcp-1.1.3        |       h7f98852_0          19 KB  conda-forge\n",
            "    xorg-libxext-1.3.4         |       h0b41bf4_2          49 KB  conda-forge\n",
            "    xorg-libxfixes-5.0.3       |    h7f98852_1004          18 KB  conda-forge\n",
            "    xorg-libxi-1.7.10          |       h7f98852_0          46 KB  conda-forge\n",
            "    xorg-libxrender-0.9.11     |       hd590300_0          37 KB  conda-forge\n",
            "    xorg-libxtst-1.2.3         |    h7f98852_1002          31 KB  conda-forge\n",
            "    xorg-recordproto-1.14.2    |    h7f98852_1002           8 KB  conda-forge\n",
            "    xorg-renderproto-0.11.1    |    h7f98852_1002           9 KB  conda-forge\n",
            "    xorg-xextproto-7.3.0       |    h0b41bf4_1003          30 KB  conda-forge\n",
            "    xorg-xproto-7.0.31         |    h7f98852_1007          73 KB  conda-forge\n",
            "    xxhash-0.8.2               |       hd590300_0          95 KB  conda-forge\n",
            "    yaml-0.2.5                 |       h7f98852_2          87 KB  conda-forge\n",
            "    yarl-1.9.4                 |  py310h2372a71_0         111 KB  conda-forge\n",
            "    zlib-1.2.13                |       h4ab18f5_6          91 KB  conda-forge\n",
            "    zstandard-0.23.0           |  py310h64cae3c_0         398 KB  conda-forge\n",
            "    zstd-1.5.6                 |       ha6fb4c9_0         542 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        2.73 GB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  about-time         conda-forge/noarch::about-time-4.2.1-pyhd8ed1ab_0 \n",
            "  aiohttp            conda-forge/linux-64::aiohttp-3.9.5-py310h2372a71_0 \n",
            "  aiosignal          conda-forge/noarch::aiosignal-1.3.1-pyhd8ed1ab_0 \n",
            "  alive-progress     conda-forge/noarch::alive-progress-3.1.5-pyhd8ed1ab_0 \n",
            "  alsa-lib           conda-forge/linux-64::alsa-lib-1.2.6.1-h7f98852_0 \n",
            "  aragorn            bioconda/linux-64::aragorn-1.2.41-h031d066_3 \n",
            "  aria2              conda-forge/linux-64::aria2-1.36.0-h1e4e653_3 \n",
            "  async-timeout      conda-forge/noarch::async-timeout-4.0.3-pyhd8ed1ab_0 \n",
            "  attrs              conda-forge/noarch::attrs-23.2.0-pyh71513ae_0 \n",
            "  aws-c-auth         conda-forge/linux-64::aws-c-auth-0.7.20-h5f1c8d9_0 \n",
            "  aws-c-cal          conda-forge/linux-64::aws-c-cal-0.6.12-h2ba76a8_0 \n",
            "  aws-c-common       conda-forge/linux-64::aws-c-common-0.9.17-h4ab18f5_0 \n",
            "  aws-c-compression  conda-forge/linux-64::aws-c-compression-0.2.18-h36a0aea_4 \n",
            "  aws-c-event-stream conda-forge/linux-64::aws-c-event-stream-0.4.2-h161de36_10 \n",
            "  aws-c-http         conda-forge/linux-64::aws-c-http-0.8.1-h63f54a0_13 \n",
            "  aws-c-io           conda-forge/linux-64::aws-c-io-0.14.8-h96d4d28_0 \n",
            "  aws-c-mqtt         conda-forge/linux-64::aws-c-mqtt-0.10.4-hcc7299c_2 \n",
            "  aws-c-s3           conda-forge/linux-64::aws-c-s3-0.5.9-h10bd90f_0 \n",
            "  aws-c-sdkutils     conda-forge/linux-64::aws-c-sdkutils-0.1.16-h36a0aea_0 \n",
            "  aws-checksums      conda-forge/linux-64::aws-checksums-0.1.18-h36a0aea_4 \n",
            "  aws-crt-cpp        conda-forge/linux-64::aws-crt-cpp-0.26.8-h4f3a3cc_11 \n",
            "  aws-sdk-cpp        conda-forge/linux-64::aws-sdk-cpp-1.11.329-h51dfee4_0 \n",
            "  backports          conda-forge/noarch::backports-1.0-pyhd8ed1ab_3 \n",
            "  backports.tempfile conda-forge/noarch::backports.tempfile-1.0-py_0 \n",
            "  backports.weakref  conda-forge/noarch::backports.weakref-1.0.post1-pyhd8ed1ab_1003 \n",
            "  bcbio-gff          bioconda/noarch::bcbio-gff-0.7.1-pyh7cba7a3_1 \n",
            "  biopython          conda-forge/linux-64::biopython-1.81-py310h2372a71_1 \n",
            "  black              conda-forge/linux-64::black-24.4.2-py310hff52083_0 \n",
            "  blast              bioconda/linux-64::blast-2.15.0-pl5321h6f7f691_1 \n",
            "  brotli             conda-forge/linux-64::brotli-1.1.0-hd590300_1 \n",
            "  brotli-bin         conda-forge/linux-64::brotli-bin-1.1.0-hd590300_1 \n",
            "  bx-python          bioconda/linux-64::bx-python-0.12.0-py310ha025fb0_0 \n",
            "  cached-property    conda-forge/noarch::cached-property-1.5.2-hd8ed1ab_1 \n",
            "  cached_property    conda-forge/noarch::cached_property-1.5.2-pyha770c72_1 \n",
            "  cairo              pkgs/main/linux-64::cairo-1.16.0-hb05425b_5 \n",
            "  capnproto          conda-forge/linux-64::capnproto-1.0.2-h2b92303_1 \n",
            "  click              conda-forge/noarch::click-8.1.7-unix_pyh707e725_0 \n",
            "  colorama           conda-forge/noarch::colorama-0.4.6-pyhd8ed1ab_0 \n",
            "  contourpy          conda-forge/linux-64::contourpy-1.2.1-py310hd41b1e2_0 \n",
            "  coverage           conda-forge/linux-64::coverage-7.6.0-py310h5b4e0ec_0 \n",
            "  cuda-cudart        conda-forge/linux-64::cuda-cudart-12.5.82-he02047a_0 \n",
            "  cuda-cudart_linux~ conda-forge/noarch::cuda-cudart_linux-64-12.5.82-h85509e4_0 \n",
            "  cuda-nvrtc         conda-forge/linux-64::cuda-nvrtc-12.5.82-he02047a_0 \n",
            "  cuda-nvtx          conda-forge/linux-64::cuda-nvtx-12.5.82-he02047a_0 \n",
            "  cuda-version       conda-forge/noarch::cuda-version-12.5-hd4f0392_3 \n",
            "  cudnn              conda-forge/linux-64::cudnn-8.9.7.29-h092f7fd_3 \n",
            "  curl               pkgs/main/linux-64::curl-8.7.1-hdbd6064_0 \n",
            "  cycler             conda-forge/noarch::cycler-0.12.1-pyhd8ed1ab_0 \n",
            "  datasets           conda-forge/noarch::datasets-2.20.0-pyhd8ed1ab_0 \n",
            "  dill               conda-forge/noarch::dill-0.3.8-pyhd8ed1ab_0 \n",
            "  dnaapler           bioconda/noarch::dnaapler-0.7.0-pyhdfd78af_0 \n",
            "  entrez-direct      bioconda/linux-64::entrez-direct-22.1-he881be0_0 \n",
            "  exceptiongroup     conda-forge/noarch::exceptiongroup-1.2.2-pyhd8ed1ab_0 \n",
            "  fastpath           conda-forge/linux-64::fastpath-1.9-py310h2372a71_3 \n",
            "  filelock           conda-forge/noarch::filelock-3.15.4-pyhd8ed1ab_0 \n",
            "  foldseek           bioconda/linux-64::foldseek-9.427df8a-pl5321hb365157_1 \n",
            "  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0 \n",
            "  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0 \n",
            "  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0 \n",
            "  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-h77eed37_2 \n",
            "  fontconfig         conda-forge/linux-64::fontconfig-2.14.2-h14ed4e7_0 \n",
            "  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0 \n",
            "  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-0 \n",
            "  fonttools          conda-forge/linux-64::fonttools-4.53.1-py310h5b4e0ec_0 \n",
            "  freetype           conda-forge/linux-64::freetype-2.12.1-h267a509_2 \n",
            "  frozenlist         conda-forge/linux-64::frozenlist-1.4.1-py310h2372a71_0 \n",
            "  fsspec             conda-forge/noarch::fsspec-2024.5.0-pyhff2d567_0 \n",
            "  gawk               conda-forge/linux-64::gawk-5.3.0-ha916aea_0 \n",
            "  gettext            conda-forge/linux-64::gettext-0.22.5-h59595ed_2 \n",
            "  gettext-tools      conda-forge/linux-64::gettext-tools-0.22.5-h59595ed_2 \n",
            "  gflags             conda-forge/linux-64::gflags-2.2.2-he1b5a44_1004 \n",
            "  giflib             conda-forge/linux-64::giflib-5.2.2-hd590300_0 \n",
            "  glib               conda-forge/linux-64::glib-2.78.4-hfc55251_0 \n",
            "  glib-tools         conda-forge/linux-64::glib-tools-2.78.4-hfc55251_0 \n",
            "  glog               conda-forge/linux-64::glog-0.7.1-hbabe93e_0 \n",
            "  gmp                conda-forge/linux-64::gmp-6.3.0-hac33072_2 \n",
            "  gmpy2              conda-forge/linux-64::gmpy2-2.1.5-py310hc7909c9_1 \n",
            "  grapheme           conda-forge/noarch::grapheme-0.6.0-pyhd8ed1ab_0 \n",
            "  graphite2          pkgs/main/linux-64::graphite2-1.3.14-h295c915_1 \n",
            "  gsl                conda-forge/linux-64::gsl-2.7-he838d99_0 \n",
            "  h2                 conda-forge/noarch::h2-4.1.0-pyhd8ed1ab_0 \n",
            "  h5py               conda-forge/linux-64::h5py-3.11.0-nompi_py310hf054cd7_102 \n",
            "  harfbuzz           pkgs/main/linux-64::harfbuzz-4.3.0-hf52aaf7_2 \n",
            "  hdf5               conda-forge/linux-64::hdf5-1.14.3-nompi_h4f84152_101 \n",
            "  hpack              conda-forge/noarch::hpack-4.0.0-pyh9f0ad1d_0 \n",
            "  huggingface_hub    conda-forge/noarch::huggingface_hub-0.23.4-pyhd8ed1ab_0 \n",
            "  hyperframe         conda-forge/noarch::hyperframe-6.0.1-pyhd8ed1ab_0 \n",
            "  infernal           bioconda/linux-64::infernal-1.1.5-pl5321h031d066_2 \n",
            "  iniconfig          conda-forge/noarch::iniconfig-2.0.0-pyhd8ed1ab_0 \n",
            "  isort              conda-forge/noarch::isort-5.13.2-pyhd8ed1ab_0 \n",
            "  jinja2             conda-forge/noarch::jinja2-3.1.4-pyhd8ed1ab_0 \n",
            "  jpeg               conda-forge/linux-64::jpeg-9e-h0b41bf4_3 \n",
            "  kiwisolver         conda-forge/linux-64::kiwisolver-1.4.5-py310hd41b1e2_1 \n",
            "  lcms2              conda-forge/linux-64::lcms2-2.15-hfd0df8a_0 \n",
            "  lerc               conda-forge/linux-64::lerc-4.0.0-h27087fc_0 \n",
            "  libabseil          conda-forge/linux-64::libabseil-20240116.2-cxx17_he02047a_1 \n",
            "  libaec             conda-forge/linux-64::libaec-1.1.3-h59595ed_0 \n",
            "  libarrow           conda-forge/linux-64::libarrow-16.1.0-h06228bf_5_cpu \n",
            "  libarrow-acero     conda-forge/linux-64::libarrow-acero-16.1.0-hac33072_5_cpu \n",
            "  libarrow-dataset   conda-forge/linux-64::libarrow-dataset-16.1.0-hac33072_5_cpu \n",
            "  libarrow-substrait conda-forge/linux-64::libarrow-substrait-16.1.0-h7e0c224_5_cpu \n",
            "  libasprintf        conda-forge/linux-64::libasprintf-0.22.5-h661eb56_2 \n",
            "  libasprintf-devel  conda-forge/linux-64::libasprintf-devel-0.22.5-h661eb56_2 \n",
            "  libblas            conda-forge/linux-64::libblas-3.9.0-22_linux64_openblas \n",
            "  libbrotlicommon    conda-forge/linux-64::libbrotlicommon-1.1.0-hd590300_1 \n",
            "  libbrotlidec       conda-forge/linux-64::libbrotlidec-1.1.0-hd590300_1 \n",
            "  libbrotlienc       conda-forge/linux-64::libbrotlienc-1.1.0-hd590300_1 \n",
            "  libcblas           conda-forge/linux-64::libcblas-3.9.0-22_linux64_openblas \n",
            "  libcrc32c          conda-forge/linux-64::libcrc32c-1.1.2-h9c3ff4c_0 \n",
            "  libcublas          conda-forge/linux-64::libcublas-12.5.3.2-he02047a_0 \n",
            "  libcufft           conda-forge/linux-64::libcufft-11.2.3.61-he02047a_0 \n",
            "  libcups            conda-forge/linux-64::libcups-2.3.3-h36d4200_3 \n",
            "  libcurand          conda-forge/linux-64::libcurand-10.3.6.82-he02047a_0 \n",
            "  libcusolver        conda-forge/linux-64::libcusolver-11.6.3.83-he02047a_0 \n",
            "  libcusparse        conda-forge/linux-64::libcusparse-12.5.1.3-he02047a_0 \n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.17-h0b41bf4_0 \n",
            "  libevent           conda-forge/linux-64::libevent-2.1.12-hf998b51_1 \n",
            "  libgettextpo       conda-forge/linux-64::libgettextpo-0.22.5-h59595ed_2 \n",
            "  libgettextpo-devel conda-forge/linux-64::libgettextpo-devel-0.22.5-h59595ed_2 \n",
            "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-14.1.0-h69a702a_0 \n",
            "  libgfortran5       conda-forge/linux-64::libgfortran5-14.1.0-hc5f4f2c_0 \n",
            "  libglib            conda-forge/linux-64::libglib-2.78.4-h783c2da_0 \n",
            "  libgoogle-cloud    conda-forge/linux-64::libgoogle-cloud-2.24.0-h2736e30_0 \n",
            "  libgoogle-cloud-s~ conda-forge/linux-64::libgoogle-cloud-storage-2.24.0-h3d9a0c8_0 \n",
            "  libgrpc            conda-forge/linux-64::libgrpc-1.62.2-h15f2491_0 \n",
            "  libhwloc           conda-forge/linux-64::libhwloc-2.9.1-hd6dc26d_0 \n",
            "  libiconv           conda-forge/linux-64::libiconv-1.17-hd590300_2 \n",
            "  libidn2            conda-forge/linux-64::libidn2-2.3.7-hd590300_0 \n",
            "  liblapack          conda-forge/linux-64::liblapack-3.9.0-22_linux64_openblas \n",
            "  libmagma           conda-forge/linux-64::libmagma-2.7.2-h173bb3b_2 \n",
            "  libmagma_sparse    conda-forge/linux-64::libmagma_sparse-2.7.2-h173bb3b_3 \n",
            "  libnsl             conda-forge/linux-64::libnsl-2.0.1-hd590300_0 \n",
            "  libnvjitlink       conda-forge/linux-64::libnvjitlink-12.5.82-he02047a_0 \n",
            "  libopenblas        conda-forge/linux-64::libopenblas-0.3.27-pthreads_hac2b453_1 \n",
            "  libparquet         conda-forge/linux-64::libparquet-16.1.0-h6a7eafb_5_cpu \n",
            "  libpng             conda-forge/linux-64::libpng-1.6.43-h2797004_0 \n",
            "  libprotobuf        conda-forge/linux-64::libprotobuf-4.25.3-h08a7969_0 \n",
            "  libre2-11          conda-forge/linux-64::libre2-11-2023.09.01-h5a48ba9_2 \n",
            "  libsentencepiece   conda-forge/linux-64::libsentencepiece-0.2.0-he81a138_2 \n",
            "  libsqlite          conda-forge/linux-64::libsqlite-3.46.0-hde9e2c9_0 \n",
            "  libthrift          conda-forge/linux-64::libthrift-0.19.0-hb90f79a_1 \n",
            "  libtiff            conda-forge/linux-64::libtiff-4.5.0-h6adf6a1_2 \n",
            "  libtorch           conda-forge/linux-64::libtorch-2.3.1-cuda120_h2b0da52_300 \n",
            "  libunistring       conda-forge/linux-64::libunistring-0.9.10-h7f98852_0 \n",
            "  libutf8proc        conda-forge/linux-64::libutf8proc-2.8.0-h166bdaf_0 \n",
            "  libuv              conda-forge/linux-64::libuv-1.48.0-hd590300_0 \n",
            "  libwebp-base       conda-forge/linux-64::libwebp-base-1.4.0-hd590300_0 \n",
            "  libxcb             conda-forge/linux-64::libxcb-1.16-hd590300_0 \n",
            "  libxcrypt          conda-forge/linux-64::libxcrypt-4.4.36-hd590300_1 \n",
            "  libzlib            conda-forge/linux-64::libzlib-1.2.13-h4ab18f5_6 \n",
            "  llvm-openmp        conda-forge/linux-64::llvm-openmp-18.1.7-ha31de31_0 \n",
            "  loguru             conda-forge/linux-64::loguru-0.7.2-py310hff52083_1 \n",
            "  markupsafe         conda-forge/linux-64::markupsafe-2.1.5-py310h2372a71_0 \n",
            "  mash               bioconda/linux-64::mash-2.3-hc74b729_7 \n",
            "  matplotlib-base    conda-forge/linux-64::matplotlib-base-3.9.1-py310h0b1de36_0 \n",
            "  minced             bioconda/noarch::minced-0.4.2-hdfd78af_1 \n",
            "  mkl                conda-forge/linux-64::mkl-2023.2.0-h84fe81f_50496 \n",
            "  mmseqs2            bioconda/linux-64::mmseqs2-13.45111-pl5321hf1761c0_2 \n",
            "  mpc                conda-forge/linux-64::mpc-1.3.1-hfe3b2da_0 \n",
            "  mpfr               conda-forge/linux-64::mpfr-4.2.1-h9458935_1 \n",
            "  mpmath             conda-forge/noarch::mpmath-1.3.0-pyhd8ed1ab_0 \n",
            "  multidict          conda-forge/linux-64::multidict-6.0.5-py310h2372a71_0 \n",
            "  multiprocess       conda-forge/linux-64::multiprocess-0.70.16-py310h2372a71_0 \n",
            "  munkres            conda-forge/noarch::munkres-1.1.4-pyh9f0ad1d_0 \n",
            "  mypy_extensions    conda-forge/noarch::mypy_extensions-1.0.0-pyha770c72_0 \n",
            "  ncbi-vdb           bioconda/linux-64::ncbi-vdb-3.1.1-h4ac6f70_1 \n",
            "  nccl               conda-forge/linux-64::nccl-2.22.3.1-hbc370b7_0 \n",
            "  networkx           conda-forge/noarch::networkx-3.3-pyhd8ed1ab_1 \n",
            "  numpy              conda-forge/linux-64::numpy-2.0.0-py310h515e003_0 \n",
            "  openjdk            conda-forge/linux-64::openjdk-11.0.15-hc6918da_0 \n",
            "  openjpeg           conda-forge/linux-64::openjpeg-2.5.0-hfec8fc6_2 \n",
            "  orc                conda-forge/linux-64::orc-2.0.1-h17fec99_1 \n",
            "  pandas             conda-forge/linux-64::pandas-2.2.2-py310hf9f9076_1 \n",
            "  pathspec           conda-forge/noarch::pathspec-0.12.1-pyhd8ed1ab_0 \n",
            "  patsy              conda-forge/noarch::patsy-0.5.6-pyhd8ed1ab_0 \n",
            "  pcre               conda-forge/linux-64::pcre-8.45-h9c3ff4c_0 \n",
            "  perl               conda-forge/linux-64::perl-5.32.1-7_hd590300_perl5 \n",
            "  perl-archive-tar   bioconda/noarch::perl-archive-tar-2.40-pl5321hdfd78af_0 \n",
            "  perl-carp          conda-forge/noarch::perl-carp-1.50-pl5321hd8ed1ab_0 \n",
            "  perl-common-sense  conda-forge/noarch::perl-common-sense-3.75-pl5321hd8ed1ab_0 \n",
            "  perl-compress-raw~ conda-forge/linux-64::perl-compress-raw-bzip2-2.201-pl5321h166bdaf_0 \n",
            "  perl-compress-raw~ conda-forge/linux-64::perl-compress-raw-zlib-2.202-pl5321h166bdaf_0 \n",
            "  perl-encode        conda-forge/linux-64::perl-encode-3.21-pl5321hd590300_0 \n",
            "  perl-exporter      conda-forge/noarch::perl-exporter-5.74-pl5321hd8ed1ab_0 \n",
            "  perl-exporter-tiny conda-forge/noarch::perl-exporter-tiny-1.002002-pl5321hd8ed1ab_0 \n",
            "  perl-extutils-mak~ conda-forge/noarch::perl-extutils-makemaker-7.70-pl5321hd8ed1ab_0 \n",
            "  perl-io-compress   bioconda/linux-64::perl-io-compress-2.201-pl5321hdbdd923_2 \n",
            "  perl-io-zlib       bioconda/noarch::perl-io-zlib-1.14-pl5321hdfd78af_0 \n",
            "  perl-json          bioconda/noarch::perl-json-4.10-pl5321hdfd78af_0 \n",
            "  perl-json-xs       bioconda/linux-64::perl-json-xs-2.34-pl5321h4ac6f70_6 \n",
            "  perl-list-moreuti~ bioconda/noarch::perl-list-moreutils-0.430-pl5321hdfd78af_0 \n",
            "  perl-list-moreuti~ bioconda/linux-64::perl-list-moreutils-xs-0.430-pl5321h031d066_2 \n",
            "  perl-parent        conda-forge/noarch::perl-parent-0.241-pl5321hd8ed1ab_0 \n",
            "  perl-pathtools     conda-forge/linux-64::perl-pathtools-3.75-pl5321h166bdaf_0 \n",
            "  perl-scalar-list-~ conda-forge/linux-64::perl-scalar-list-utils-1.63-pl5321h166bdaf_0 \n",
            "  perl-storable      conda-forge/linux-64::perl-storable-3.15-pl5321h166bdaf_0 \n",
            "  perl-types-serial~ bioconda/noarch::perl-types-serialiser-1.01-pl5321hdfd78af_0 \n",
            "  phanotate          bioconda/linux-64::phanotate-1.5.1-h0dbaff4_1 \n",
            "  pharokka           bioconda/noarch::pharokka-1.7.3-pyhdfd78af_0 \n",
            "  phold              bioconda/noarch::phold-0.2.0-pyhdfd78af_0 \n",
            "  pillow             pkgs/main/linux-64::pillow-10.4.0-py310h5eee18b_0 \n",
            "  pixman             conda-forge/linux-64::pixman-0.43.2-h59595ed_0 \n",
            "  psutil             conda-forge/linux-64::psutil-6.0.0-py310hc51659f_0 \n",
            "  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-h36c2ea0_1001 \n",
            "  pyarrow            conda-forge/linux-64::pyarrow-16.1.0-py310hb7f781d_2 \n",
            "  pyarrow-core       conda-forge/linux-64::pyarrow-core-16.1.0-py310hcccf80d_2_cpu \n",
            "  pyarrow-hotfix     conda-forge/noarch::pyarrow-hotfix-0.6-pyhd8ed1ab_0 \n",
            "  pycirclize         conda-forge/noarch::pycirclize-1.6.0-pyhd8ed1ab_0 \n",
            "  pyhmmer            bioconda/linux-64::pyhmmer-0.10.14-py310h7c593f9_0 \n",
            "  pyparsing          conda-forge/noarch::pyparsing-3.1.2-pyhd8ed1ab_0 \n",
            "  pyrodigal          bioconda/linux-64::pyrodigal-3.5.0-py310h7c593f9_0 \n",
            "  pyrodigal-gv       bioconda/noarch::pyrodigal-gv-0.3.1-pyh7cba7a3_0 \n",
            "  pytest             conda-forge/noarch::pytest-8.2.2-pyhd8ed1ab_0 \n",
            "  pytest-cov         conda-forge/noarch::pytest-cov-5.0.0-pyhd8ed1ab_0 \n",
            "  python-dateutil    conda-forge/noarch::python-dateutil-2.9.0-pyhd8ed1ab_0 \n",
            "  python-tzdata      conda-forge/noarch::python-tzdata-2024.1-pyhd8ed1ab_0 \n",
            "  python-xxhash      conda-forge/linux-64::python-xxhash-3.4.1-py310h2372a71_0 \n",
            "  python_abi         conda-forge/linux-64::python_abi-3.10-4_cp310 \n",
            "  pytorch            conda-forge/linux-64::pytorch-2.3.1-cuda120_py310h2c91c31_300 \n",
            "  pytz               conda-forge/noarch::pytz-2024.1-pyhd8ed1ab_0 \n",
            "  pyyaml             conda-forge/linux-64::pyyaml-6.0.1-py310h2372a71_1 \n",
            "  qhull              conda-forge/linux-64::qhull-2020.2-h434a139_5 \n",
            "  re2                conda-forge/linux-64::re2-2023.09.01-h7f4b329_2 \n",
            "  regex              conda-forge/linux-64::regex-2024.5.15-py310hc51659f_0 \n",
            "  ruamel.yaml.clib   conda-forge/linux-64::ruamel.yaml.clib-0.2.8-py310h2372a71_0 \n",
            "  s2n                conda-forge/linux-64::s2n-1.4.13-he19d79f_0 \n",
            "  safetensors        conda-forge/linux-64::safetensors-0.4.3-py310hcb5633a_0 \n",
            "  scipy              conda-forge/linux-64::scipy-1.14.0-py310h93e2701_1 \n",
            "  seaborn            conda-forge/noarch::seaborn-0.13.2-hd8ed1ab_2 \n",
            "  seaborn-base       conda-forge/noarch::seaborn-base-0.13.2-pyhd8ed1ab_2 \n",
            "  sentencepiece      conda-forge/linux-64::sentencepiece-0.2.0-hff52083_2 \n",
            "  sentencepiece-pyt~ conda-forge/linux-64::sentencepiece-python-0.2.0-py310h91407d4_2 \n",
            "  sentencepiece-spm  conda-forge/linux-64::sentencepiece-spm-0.2.0-he81a138_2 \n",
            "  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0 \n",
            "  sleef              conda-forge/linux-64::sleef-3.6.1-h3400bea_1 \n",
            "  snappy             conda-forge/linux-64::snappy-1.2.1-ha2e4443_0 \n",
            "  statsmodels        conda-forge/linux-64::statsmodels-0.14.2-py310h261611a_0 \n",
            "  sympy              conda-forge/noarch::sympy-1.13.0-pypyh2585a3b_103 \n",
            "  tbb                conda-forge/linux-64::tbb-2021.9.0-hf52228f_0 \n",
            "  textwrap3          conda-forge/noarch::textwrap3-0.9.2-py_0 \n",
            "  tokenizers         conda-forge/linux-64::tokenizers-0.19.1-py310h320607d_0 \n",
            "  toml               conda-forge/noarch::toml-0.10.2-pyhd8ed1ab_0 \n",
            "  tomli              conda-forge/noarch::tomli-2.0.1-pyhd8ed1ab_0 \n",
            "  transformers       conda-forge/noarch::transformers-4.41.2-pyhd8ed1ab_0 \n",
            "  trnascan-se        bioconda/linux-64::trnascan-se-2.0.12-pl5321h031d066_0 \n",
            "  typing-extensions  conda-forge/noarch::typing-extensions-4.12.2-hd8ed1ab_0 \n",
            "  typing_extensions  conda-forge/noarch::typing_extensions-4.12.2-pyha770c72_0 \n",
            "  unicodedata2       conda-forge/linux-64::unicodedata2-15.1.0-py310h2372a71_0 \n",
            "  wget               conda-forge/linux-64::wget-1.21.4-hda4d442_0 \n",
            "  xorg-fixesproto    conda-forge/linux-64::xorg-fixesproto-5.0-h7f98852_1002 \n",
            "  xorg-inputproto    conda-forge/linux-64::xorg-inputproto-2.3.2-h7f98852_1002 \n",
            "  xorg-kbproto       conda-forge/linux-64::xorg-kbproto-1.0.7-h7f98852_1002 \n",
            "  xorg-libx11        conda-forge/linux-64::xorg-libx11-1.8.9-hb711507_1 \n",
            "  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.11-hd590300_0 \n",
            "  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.3-h7f98852_0 \n",
            "  xorg-libxext       conda-forge/linux-64::xorg-libxext-1.3.4-h0b41bf4_2 \n",
            "  xorg-libxfixes     conda-forge/linux-64::xorg-libxfixes-5.0.3-h7f98852_1004 \n",
            "  xorg-libxi         conda-forge/linux-64::xorg-libxi-1.7.10-h7f98852_0 \n",
            "  xorg-libxrender    conda-forge/linux-64::xorg-libxrender-0.9.11-hd590300_0 \n",
            "  xorg-libxtst       conda-forge/linux-64::xorg-libxtst-1.2.3-h7f98852_1002 \n",
            "  xorg-recordproto   conda-forge/linux-64::xorg-recordproto-1.14.2-h7f98852_1002 \n",
            "  xorg-renderproto   conda-forge/linux-64::xorg-renderproto-0.11.1-h7f98852_1002 \n",
            "  xorg-xextproto     conda-forge/linux-64::xorg-xextproto-7.3.0-h0b41bf4_1003 \n",
            "  xorg-xproto        conda-forge/linux-64::xorg-xproto-7.0.31-h7f98852_1007 \n",
            "  xxhash             conda-forge/linux-64::xxhash-0.8.2-hd590300_0 \n",
            "  yaml               conda-forge/linux-64::yaml-0.2.5-h7f98852_2 \n",
            "  yarl               conda-forge/linux-64::yarl-1.9.4-py310h2372a71_0 \n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  boltons            pkgs/main/linux-64::boltons-23.0.0-py~ --> conda-forge/noarch::boltons-24.0.0-pyhd8ed1ab_0 \n",
            "  brotli-python      pkgs/main::brotli-python-1.0.9-py312h~ --> conda-forge::brotli-python-1.1.0-py310hc6cd4ac_1 \n",
            "  c-ares                pkgs/main::c-ares-1.19.1-h5eee18b_0 --> conda-forge::c-ares-1.32.2-h4bc722e_0 \n",
            "  ca-certificates    pkgs/main::ca-certificates-2024.3.11-~ --> conda-forge::ca-certificates-2024.7.4-hbcca054_0 \n",
            "  certifi            pkgs/main/linux-64::certifi-2024.6.2-~ --> conda-forge/noarch::certifi-2024.7.4-pyhd8ed1ab_0 \n",
            "  cryptography       pkgs/main::cryptography-42.0.5-py312h~ --> conda-forge::cryptography-42.0.8-py310hb1bd9d3_0 \n",
            "  frozendict         pkgs/main::frozendict-2.4.2-py312h06a~ --> conda-forge::frozendict-2.4.4-py310hc51659f_0 \n",
            "  libgcc-ng          pkgs/main::libgcc-ng-11.2.0-h1234567_1 --> conda-forge::libgcc-ng-14.1.0-h77fa898_0 \n",
            "  libgomp              pkgs/main::libgomp-11.2.0-h1234567_1 --> conda-forge::libgomp-14.1.0-h77fa898_0 \n",
            "  libstdcxx-ng       pkgs/main::libstdcxx-ng-11.2.0-h12345~ --> conda-forge::libstdcxx-ng-14.1.0-hc0a3c3a_0 \n",
            "  libuuid              pkgs/main::libuuid-1.41.5-h5eee18b_0 --> conda-forge::libuuid-2.38.1-h0b41bf4_0 \n",
            "  ncurses                 pkgs/main::ncurses-6.4-h6a678d5_0 --> conda-forge::ncurses-6.5-h59595ed_0 \n",
            "  openssl              pkgs/main::openssl-3.0.14-h5eee18b_0 --> conda-forge::openssl-3.3.1-h4bc722e_2 \n",
            "  packaging          pkgs/main/linux-64::packaging-23.2-py~ --> conda-forge/noarch::packaging-24.1-pyhd8ed1ab_0 \n",
            "  platformdirs       pkgs/main/linux-64::platformdirs-3.10~ --> conda-forge/noarch::platformdirs-4.2.2-pyhd8ed1ab_0 \n",
            "  pluggy             pkgs/main/linux-64::pluggy-1.0.0-py31~ --> conda-forge/noarch::pluggy-1.5.0-pyhd8ed1ab_0 \n",
            "  pysocks            pkgs/main/linux-64::pysocks-1.7.1-py3~ --> conda-forge/noarch::pysocks-1.7.1-pyha2e5f31_6 \n",
            "  requests           pkgs/main/linux-64::requests-2.32.2-p~ --> conda-forge/noarch::requests-2.32.3-pyhd8ed1ab_0 \n",
            "  ruamel.yaml        pkgs/main::ruamel.yaml-0.17.21-py312h~ --> conda-forge::ruamel.yaml-0.18.6-py310h2372a71_0 \n",
            "  setuptools         pkgs/main/linux-64::setuptools-69.5.1~ --> conda-forge/noarch::setuptools-70.3.0-pyhd8ed1ab_0 \n",
            "  urllib3            pkgs/main/linux-64::urllib3-2.2.2-py3~ --> conda-forge/noarch::urllib3-2.2.2-pyhd8ed1ab_1 \n",
            "  wheel              pkgs/main/linux-64::wheel-0.43.0-py31~ --> conda-forge/noarch::wheel-0.43.0-pyhd8ed1ab_1 \n",
            "  zlib                    pkgs/main::zlib-1.2.13-h5eee18b_1 --> conda-forge::zlib-1.2.13-h4ab18f5_6 \n",
            "  zstandard          pkgs/main::zstandard-0.22.0-py312h2c3~ --> conda-forge::zstandard-0.23.0-py310h64cae3c_0 \n",
            "  zstd                     pkgs/main::zstd-1.5.5-hc292b87_2 --> conda-forge::zstd-1.5.6-ha6fb4c9_0 \n",
            "\n",
            "The following packages will be SUPERSEDED by a higher-priority channel:\n",
            "\n",
            "  _libgcc_mutex           pkgs/main::_libgcc_mutex-0.1-main --> conda-forge::_libgcc_mutex-0.1-conda_forge \n",
            "  _openmp_mutex          pkgs/main::_openmp_mutex-5.1-1_gnu --> conda-forge::_openmp_mutex-4.5-2_kmp_llvm \n",
            "  cffi               pkgs/main::cffi-1.16.0-py312h5eee18b_1 --> conda-forge::cffi-1.16.0-py310h2fee648_0 \n",
            "  conda              pkgs/main::conda-24.5.0-py312h06a4308~ --> conda-forge::conda-24.5.0-py310hff52083_0 \n",
            "  conda-content-tru~ pkgs/main/linux-64::conda-content-tru~ --> conda-forge/noarch::conda-content-trust-0.2.0-pyhd8ed1ab_0 \n",
            "  conda-package-han~ pkgs/main/linux-64::conda-package-han~ --> conda-forge/noarch::conda-package-handling-2.3.0-pyh7900ff3_0 \n",
            "  conda-package-str~ pkgs/main/linux-64::conda-package-str~ --> conda-forge/noarch::conda-package-streaming-0.10.0-pyhd8ed1ab_0 \n",
            "  distro             pkgs/main/linux-64::distro-1.9.0-py31~ --> conda-forge/noarch::distro-1.9.0-pyhd8ed1ab_0 \n",
            "  idna               pkgs/main/linux-64::idna-3.7-py312h06~ --> conda-forge/noarch::idna-3.7-pyhd8ed1ab_0 \n",
            "  jsonpatch          pkgs/main/linux-64::jsonpatch-1.33-py~ --> conda-forge/noarch::jsonpatch-1.33-pyhd8ed1ab_0 \n",
            "  menuinst           pkgs/main::menuinst-2.1.1-py312h06a43~ --> conda-forge::menuinst-2.1.1-py310hff52083_0 \n",
            "  pip                pkgs/main/linux-64::pip-24.0-py312h06~ --> conda-forge/noarch::pip-24.0-pyhd8ed1ab_0 \n",
            "  pybind11-abi         pkgs/main::pybind11-abi-5-hd3eb1b0_0 --> conda-forge::pybind11-abi-4-hd8ed1ab_3 \n",
            "  pycosat            pkgs/main::pycosat-0.6.6-py312h5eee18~ --> conda-forge::pycosat-0.6.6-py310h2372a71_0 \n",
            "  python                pkgs/main::python-3.12.4-h5148396_1 --> conda-forge::python-3.10.14-hd12c33a_0_cpython \n",
            "  tqdm               pkgs/main/linux-64::tqdm-4.66.4-py312~ --> conda-forge/noarch::tqdm-4.66.4-pyhd8ed1ab_0 \n",
            "  truststore         pkgs/main/linux-64::truststore-0.8.0-~ --> conda-forge/noarch::truststore-0.8.0-pyhd8ed1ab_0 \n",
            "\n",
            "The following packages will be DOWNGRADED:\n",
            "\n",
            "  anaconda-anon-usa~                0.4.4-py312hfc0e8ea_100 --> 0.4.4-py310hfc0e8ea_100 \n",
            "  libmambapy                          1.5.8-py312h2dafd23_2 --> 1.5.8-py310h2dafd23_2 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages: ...working... done\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... By downloading and using the cuDNN conda packages, you accept the terms and conditions of the NVIDIA cuDNN EULA -\n",
            "  https://docs.nvidia.com/deeplearning/cudnn/sla/index.html\n",
            "\n",
            "done\n"
          ]
        }
      ],
      "source": [
        "#@title 1. Install pharokka and phold\n",
        "\n",
        "#@markdown This cell installs pharokka and phold. It will take a few minutes. Please be patient\n",
        "\n",
        "%%bash\n",
        "\n",
        "set -e\n",
        "\n",
        "PYTHON_VERSION=\"3.10\"\n",
        "PHAROKKA_VERSION=\"1.7.3\"\n",
        "PHOLD_VERSION=\"0.2.0\"\n",
        "\n",
        "echo \"python version ${PYTHON_VERSION}\"\n",
        "\n",
        "if [ ! -f CONDA_READY ]; then\n",
        "  echo \"installing python\"\n",
        "  wget -qnc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
        "  bash Miniconda3-latest-Linux-x86_64.sh -bfp /usr/local 2>&1 1>/dev/null\n",
        "  rm Miniconda3-latest-Linux-x86_64.sh\n",
        "  conda config --set auto_update_conda false\n",
        "  touch CONDA_READY\n",
        "fi\n",
        "\n",
        "if [ ! -f PHAROKKA_PHOLD_READY ]; then\n",
        "  echo \"installing pharokka and phold\"\n",
        "  conda install -y -c conda-forge -c bioconda pip pharokka==${PHAROKKA_VERSION} python=${PYTHON_VERSION} phold==${PHOLD_VERSION} pytorch=*=cuda*\n",
        "  touch PHAROKKA_PHOLD_READY\n",
        "fi\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tfltpbZ_QLfZ",
        "outputId": "1cf78eef-ccbf-4a36-833f-8c675f9f3bf3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading pharokka database. This will take a few minutes. Please be patient :)\n",
            "Downloading phold database. This will take a few minutes. Please be patient :)\n",
            "CPU times: user 899 ms, sys: 117 ms, total: 1.02 s\n",
            "Wall time: 4min 34s\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0"
            ]
          },
          "metadata": {},
          "execution_count": 2
        }
      ],
      "source": [
        "#@title 2. Download pharokka phold databases\n",
        "\n",
        "#@markdown This cell downloads the pharokka then the phold database. It will take some time (5-10 minutes probably). Please be patient.\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "print(\"Downloading pharokka database. This will take a few minutes. Please be patient :)\")\n",
        "os.system(\"install_databases.py -o pharokka_db\")\n",
        "print(\"Downloading phold database. This will take a few minutes. Please be patient :)\")\n",
        "os.system(\"phold install -d phold_db\")\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Sjdpu6R-Kig9",
        "outputId": "573a158f-5578-4d7c-b309-57240101b042"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Input file NC_043029.fasta exists\n",
            "Running pharokka\n",
            "pharokka completed successfully.\n",
            "Your output is in output_pharokka.\n",
            "Zipping the output directory so you can download it all in one go.\n",
            "Output directory has been zipped to output_pharokka.zip\n",
            "CPU times: user 130 ms, sys: 12.7 ms, total: 143 ms\n",
            "Wall time: 30 s\n"
          ]
        }
      ],
      "source": [
        "#@title 3. Run Pharokka\n",
        "\n",
        "#@markdown First, upload your phage(s) as a nucleotide input FASTA file\n",
        "\n",
        "#@markdown Click on the folder icon to the left and use the file upload button.\n",
        "\n",
        "#@markdown Once it is uploaded, write the file name in the INPUT_FILE field on the right.\n",
        "\n",
        "#@markdown Then provide a directory for pharokka's output using PHAROKKA_OUT_DIR.\n",
        "#@markdown The default is 'output_pharokka'.\n",
        "\n",
        "#@markdown Then type in a gene prediction tool for pharokka.\n",
        "#@markdown Please choose either 'phanotate', 'prodigal', or 'prodigal-gv'.\n",
        "\n",
        "#@markdown You can also provide a prefix for your output files with PHAROKKA_PREFIX.\n",
        "#@markdown If you provide nothing it will default to 'pharokka'.\n",
        "\n",
        "#@markdown You can also provide a locus tag for your output files.\n",
        "#@markdown If you provide nothing it will generate a random locus tag.\n",
        "\n",
        "#@markdown You can click FAST to turn off --fast.\n",
        "#@markdown By default it is True so that Pharokka runs faster in the Colab environment.\n",
        "\n",
        "#@markdown You can click META to turn on --meta if you have multiple phages in your input.\n",
        "\n",
        "#@markdown You can click META_HMM to turn on --meta_hmm.\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory.\n",
        "#@markdown This may be useful if your earlier pharokka run has crashed for whatever reason.\n",
        "\n",
        "#@markdown The results of Pharokka will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is PHAROKKA_OUT_DIR.zip, where PHAROKKA_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "import sys\n",
        "import subprocess\n",
        "import zipfile\n",
        "INPUT_FILE = '' #@param {type:\"string\"}\n",
        "\n",
        "if os.path.exists(INPUT_FILE):\n",
        "    print(f\"Input file {INPUT_FILE} exists\")\n",
        "else:\n",
        "    print(f\"Error: File {INPUT_FILE} does not exist\")\n",
        "    print(f\"Please check the spelling and that you have uploaded it correctly\")\n",
        "    sys.exit(1)\n",
        "\n",
        "PHAROKKA_OUT_DIR = 'output_pharokka'  #@param {type:\"string\"}\n",
        "GENE_PREDICTOR = 'phanotate'  #@param {type:\"string\"}\n",
        "allowed_gene_predictors = ['phanotate', 'prodigal', 'prodigal-gv']\n",
        "# Check if the input parameter is valid\n",
        "if GENE_PREDICTOR.lower() not in allowed_gene_predictors:\n",
        "    raise ValueError(\"Invalid GENE_PREDICTOR. Please choose from: 'phanotate', 'prodigal', 'prodigal-gv'.\")\n",
        "\n",
        "PHAROKKA_PREFIX = 'pharokka'  #@param {type:\"string\"}\n",
        "LOCUS_TAG = 'Default'  #@param {type:\"string\"}\n",
        "FAST = True  #@param {type:\"boolean\"}\n",
        "META = False  #@param {type:\"boolean\"}\n",
        "META_HMM = False  #@param {type:\"boolean\"}\n",
        "FORCE = True  #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "# Construct the command\n",
        "command = f\"pharokka.py -d pharokka_db -i {INPUT_FILE} -t 4 -o {PHAROKKA_OUT_DIR} -p {PHAROKKA_PREFIX} -l {LOCUS_TAG} -g {GENE_PREDICTOR}\"\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "\n",
        "if FAST is True:\n",
        "  command = f\"{command} --fast\"\n",
        "\n",
        "if META is True:\n",
        "  command = f\"{command} -m\"\n",
        "\n",
        "if META_HMM is True:\n",
        "  command = f\"{command} --meta_hmm\"\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running pharokka\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"pharokka completed successfully.\")\n",
        "    print(f\"Your output is in {PHAROKKA_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{PHAROKKA_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(PHAROKKA_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), PHAROKKA_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9QfjP3q-Q04f",
        "outputId": "a033f8e9-050d-4630-ab8e-f9ac9fba8000"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Running phold\n",
            "phold completed successfully.\n",
            "Your output is in output_phold.\n",
            "Zipping the output directory so you can download it all in one go.\n",
            "Output directory has been zipped to output_phold.zip\n",
            "CPU times: user 348 ms, sys: 46.2 ms, total: 394 ms\n",
            "Wall time: 1min 30s\n"
          ]
        }
      ],
      "source": [
        "#@title 4. Run phold\n",
        "\n",
        "#@markdown This cell will run phold on the output of cell 3's Pharokka run\n",
        "\n",
        "#@markdown You do not need to provide any further input files\n",
        "\n",
        "#@markdown You can now provide a directory for phold's output with PHOLD_OUT_DIR.\n",
        "#@markdown The default is 'output_phold'.\n",
        "\n",
        "#@markdown You can also provide a prefix for your output files with PHOLD_PREFIX.\n",
        "#@markdown If you provide nothing it will default to 'phold'.\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory with .\n",
        "#@markdown This may be useful if your earlier phold run has crashed for whatever reason.\n",
        "\n",
        "#@markdown If your input has multiple phages, you can click SEPARATE.\n",
        "#@markdown This will output separate GenBank files in the output directory.\n",
        "\n",
        "#@markdown The results of Phold will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is PHOLD_OUT_DIR.zip, where PHOLD_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "import subprocess\n",
        "import zipfile\n",
        "\n",
        "# phold input is pharokka output\n",
        "PHOLD_INPUT = f\"{PHAROKKA_OUT_DIR}/{PHAROKKA_PREFIX}.gbk\"\n",
        "PHOLD_OUT_DIR = 'output_phold'  #@param {type:\"string\"}\n",
        "PHOLD_PREFIX = 'phold'  #@param {type:\"string\"}\n",
        "FORCE = True  #@param {type:\"boolean\"}\n",
        "SEPARATE = False  #@param {type:\"boolean\"}\n",
        "\n",
        "# Construct the command\n",
        "command = f\"phold run -i {PHOLD_INPUT} -t 4 -o {PHOLD_OUT_DIR} -p {PHOLD_PREFIX} -d phold_db\"\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "if SEPARATE is True:\n",
        "  command = f\"{command} --separate\"\n",
        "\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running phold\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"phold completed successfully.\")\n",
        "    print(f\"Your output is in {PHOLD_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{PHOLD_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(PHOLD_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), PHOLD_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5. Install phynteny\n",
        "\n",
        "#@markdown This cell installs phynteny and downloads the models. It will take a few minutes. Please be patient\n",
        "%%bash\n",
        "PHYNTENY_VERSION=\"0.1.13\"\n",
        "NUMPY_VERSION=\"1.26.4\"\n",
        "\n",
        "if [ ! -f PHYNTENY_READY ]; then\n",
        "  echo \"installing phynteny\"\n",
        "  pip install phynteny==${PHYNTENY_VERSION} numpy==${NUMPY_VERSION}\n",
        "  echo \"Downloading phynteny models\"\n",
        "  install_models -o phynteny_models\n",
        "  touch PHYNTENY_READY\n",
        "fi\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1eUYA415PNRX",
        "outputId": "270c6d69-4cd6-4ce8-fc21-1805701a1abb"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "installing phynteny\n",
            "Requirement already satisfied: phynteny==0.1.13 in /usr/local/lib/python3.10/site-packages (0.1.13)\n",
            "Collecting numpy==1.26.4\n",
            "  Downloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)\n",
            "      61.0/61.0 kB 4.2 MB/s eta 0:00:00\n",
            "Requirement already satisfied: biopython>=1.79 in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (1.81)\n",
            "Requirement already satisfied: scikit-learn<=1.2.2 in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (1.2.2)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (2.2.2)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (8.1.7)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (1.4.2)\n",
            "Requirement already satisfied: loguru in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (0.7.2)\n",
            "Requirement already satisfied: tensorflow==2.9.1 in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (2.9.1)\n",
            "Requirement already satisfied: alive-progress>=3.0.1 in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (3.1.5)\n",
            "Requirement already satisfied: requests>=2.25.1 in /usr/local/lib/python3.10/site-packages (from phynteny==0.1.13) (2.32.3)\n",
            "Requirement already satisfied: absl-py>=1.0.0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (2.1.0)\n",
            "Requirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (1.6.3)\n",
            "Requirement already satisfied: flatbuffers<2,>=1.12 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (1.12)\n",
            "Requirement already satisfied: gast<=0.4.0,>=0.2.1 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (0.4.0)\n",
            "Requirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (0.2.0)\n",
            "Requirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (1.64.1)\n",
            "Requirement already satisfied: h5py>=2.9.0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (3.11.0)\n",
            "Requirement already satisfied: keras<2.10.0,>=2.9.0rc0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (2.9.0)\n",
            "Requirement already satisfied: keras-preprocessing>=1.1.1 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (1.1.2)\n",
            "Requirement already satisfied: libclang>=13.0.0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (18.1.1)\n",
            "Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (3.3.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (24.1)\n",
            "Requirement already satisfied: protobuf<3.20,>=3.9.2 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (3.19.6)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (70.3.0)\n",
            "Requirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (1.16.0)\n",
            "Requirement already satisfied: tensorboard<2.10,>=2.9 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (2.9.1)\n",
            "Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (0.37.1)\n",
            "Requirement already satisfied: tensorflow-estimator<2.10.0,>=2.9.0rc0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (2.9.0)\n",
            "Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (2.4.0)\n",
            "Requirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (4.12.2)\n",
            "Requirement already satisfied: wrapt>=1.11.0 in /usr/local/lib/python3.10/site-packages (from tensorflow==2.9.1->phynteny==0.1.13) (1.16.0)\n",
            "Requirement already satisfied: about-time==4.2.1 in /usr/local/lib/python3.10/site-packages (from alive-progress>=3.0.1->phynteny==0.1.13) (4.2.1)\n",
            "Requirement already satisfied: grapheme==0.6.0 in /usr/local/lib/python3.10/site-packages (from alive-progress>=3.0.1->phynteny==0.1.13) (0.6.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/site-packages (from requests>=2.25.1->phynteny==0.1.13) (2.0.4)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/site-packages (from requests>=2.25.1->phynteny==0.1.13) (3.7)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/site-packages (from requests>=2.25.1->phynteny==0.1.13) (2.2.2)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/site-packages (from requests>=2.25.1->phynteny==0.1.13) (2024.7.4)\n",
            "Requirement already satisfied: scipy>=1.3.2 in /usr/local/lib/python3.10/site-packages (from scikit-learn<=1.2.2->phynteny==0.1.13) (1.14.0)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/site-packages (from scikit-learn<=1.2.2->phynteny==0.1.13) (3.5.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/site-packages (from pandas->phynteny==0.1.13) (2.9.0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/site-packages (from pandas->phynteny==0.1.13) (2024.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/site-packages (from pandas->phynteny==0.1.13) (2024.1)\n",
            "Requirement already satisfied: wheel<1.0,>=0.23.0 in /usr/local/lib/python3.10/site-packages (from astunparse>=1.6.0->tensorflow==2.9.1->phynteny==0.1.13) (0.43.0)\n",
            "Requirement already satisfied: google-auth<3,>=1.6.3 in /usr/local/lib/python3.10/site-packages (from tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (2.32.0)\n",
            "Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /usr/local/lib/python3.10/site-packages (from tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (0.4.6)\n",
            "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.10/site-packages (from tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (3.6)\n",
            "Requirement already satisfied: tensorboard-data-server<0.7.0,>=0.6.0 in /usr/local/lib/python3.10/site-packages (from tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (0.6.1)\n",
            "Requirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /usr/local/lib/python3.10/site-packages (from tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (1.8.1)\n",
            "Requirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.10/site-packages (from tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (3.0.3)\n",
            "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /usr/local/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (5.4.0)\n",
            "Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (0.4.0)\n",
            "Requirement already satisfied: rsa<5,>=3.1.4 in /usr/local/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (4.9)\n",
            "Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.10/site-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (2.0.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.1.1 in /usr/local/lib/python3.10/site-packages (from werkzeug>=1.0.1->tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (2.1.5)\n",
            "Requirement already satisfied: pyasn1<0.7.0,>=0.4.6 in /usr/local/lib/python3.10/site-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (0.6.0)\n",
            "Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.10/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard<2.10,>=2.9->tensorflow==2.9.1->phynteny==0.1.13) (3.2.2)\n",
            "Downloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)\n",
            "    18.2/18.2 MB 98.7 MB/s eta 0:00:00\n",
            "Installing collected packages: numpy\n",
            "  Attempting uninstall: numpy\n",
            "    Found existing installation: numpy 2.0.0\n",
            "    Uninstalling numpy-2.0.0:\n",
            "      Successfully uninstalled numpy-2.0.0\n",
            "Successfully installed numpy-1.26.4\n",
            "Downloading phynteny models\n",
            "Downloading Phynteny models to phynteny_models\n",
            "All phynteny models have already been downloaded and checked.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {
        "id": "BlPLP3vlhWYs",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6186a5db-e111-483d-9254-15b210c24096"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Running phynteny\n",
            "phynteny completed successfully.\n",
            "Your output is in output_phynteny.\n",
            "Zipping the output directory so you can download it all in one go.\n",
            "Output directory has been zipped to output_phynteny.zip\n",
            "CPU times: user 202 ms, sys: 18.8 ms, total: 221 ms\n",
            "Wall time: 45.3 s\n"
          ]
        }
      ],
      "source": [
        "#@title 6. Run Phynteny\n",
        "\n",
        "#@markdown This cell will run phynteny on the output of cell 4's Phold run to predict the function of remaining hypothetical proteins\n",
        "\n",
        "#@markdown You do not need to provide any further input files\n",
        "\n",
        "#@markdown You can now provide a directory for phynteny's output with PHYNTENY_OUT_DIR.\n",
        "#@markdown The default is 'output_phynteny'.\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory with .\n",
        "#@markdown This may be useful if your phynteny run has crashed for whatever reason.\n",
        "\n",
        "#@markdown The results of Phynteny will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is PHYNTENY_OUT_DIR.zip, where PHYNTENY_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "import subprocess\n",
        "import zipfile\n",
        "\n",
        "# phynteny input is pharokka output\n",
        "PHYNTENY_INPUT = f\"{PHOLD_OUT_DIR}/{PHOLD_PREFIX}.gbk\"\n",
        "PHYNTENY_OUT_DIR = 'output_phynteny'  #@param {type:\"string\"}\n",
        "FORCE = False  #@param {type:\"boolean\"}\n",
        "\n",
        "# Construct the command\n",
        "command = f\"phynteny {PHYNTENY_INPUT} -m phynteny_models -o {PHYNTENY_OUT_DIR}\"\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running phynteny\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"phynteny completed successfully.\")\n",
        "    print(f\"Your output is in {PHYNTENY_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{PHYNTENY_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(PHYNTENY_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), PHYNTENY_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "authorship_tag": "ABX9TyPhV6L/1JdS3GUiG5UFg6cx",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "nbformat": 4,
  "nbformat_minor": 0
}