{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gbouras13/pharokka/blob/master/run_pharokka_and_phold_and_phynteny.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QGd2GEI3N-02"
      },
      "source": [
        "#Pharokka + Phold + Phynteny\n",
        "\n",
        "[pharokka](https://github.com/gbouras13/pharokka) is a rapid standardised annotation tool for bacteriophage genomes and metagenomes. You can read more about pharokka in the [documentation](https://pharokka.readthedocs.io/).\n",
        "\n",
        "[phold](https://github.com/gbouras13/phold) is a sensitive annotation tool for bacteriophage genomes and metagenomes using protein structural homology. You can read more about phold in the [documentation](https://phold.readthedocs.io/).\n",
        "\n",
        "phold uses the [ProstT5](https://github.com/mheinzinger/ProstT5) protein language model to translate protein amino acid sequences to the 3Di token alphabet used by [Foldseek](https://github.com/steineggerlab/foldseek). Foldseek is then used to search these against a database of 803k protein structures mostly predicted using [Colabfold](https://github.com/sokrypton/ColabFold).\n",
        "\n",
        "[phyntney](https://github.com/susiegriggo/Phynteny_transformer) uses phage synteny (the conserved gene order across phages) with a hybrid transformer/LSTM architecture to assign hypothetical phage proteins to a PHROG category.\n",
        "\n",
        "The tools are best run sequentially, as Pharokka conducts extra annotation steps like tRNA, tmRNA, CRISPR and INPHARED searches that Phold lacks (for now at least). Pharokka will also (rarely) annotate CDS that Phold can miss. Phynteny can then help annotate remaining hypothetical proteins with a PHROG category.\n",
        "\n",
        "* **Before you start, please make sure you change the runtime to T4 GPU (or any other kind of GPU if you have $$$), otherwise Phold won't be installed properly**\n",
        "* To do this, go to the top toolbar, then to Runtime -> Change runtime type -> Hardware accelerator\n",
        "\n",
        "* To run the cells, press the play button on the left side\n",
        "* Cells 1, 2 and 3 install pharokka, phold and phyntney and download the databases/models.\n",
        "* Once they have been run, you can re-run Cell 4 (to run Pharokka), Cell 5 (to run Phold) and Cell 6 (to run Phynteny) as many times as you would like\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ii39RG8eOZUx"
      },
      "outputs": [],
      "source": [
        "\n",
        "#@title 1. Install miniforge\n",
        "\n",
        "#@markdown This cell installs miniforge\n",
        "\n",
        "%%bash\n",
        "\n",
        "set -e\n",
        "\n",
        "PYTHON_VERSION=$(python3 -c \"import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')\")\n",
        "\n",
        "echo \"python version ${PYTHON_VERSION}\"\n",
        "\n",
        "if [ ! -f CONDA_READY ]; then\n",
        "  echo \"installing miniforge\"\n",
        "\n",
        "  # miniforge 25.9.1 introduces some issue - latest as of 7 Nov 2025 - see https://github.com/gbouras13/phold/issues/106\n",
        "  # issue is fixed if you use the previous release (25.3.1)\n",
        "  #wget -qnc https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh\n",
        "  wget -qnc https://github.com/conda-forge/miniforge/releases/download/25.3.1-0/Miniforge3-25.3.1-0-Linux-x86_64.sh\n",
        "  bash Miniforge3-25.3.1-0-Linux-x86_64.sh -bfp /usr/local 2>&1 1>/dev/null\n",
        "  conda config --set auto_update_conda false\n",
        "  touch CONDA_READY\n",
        "fi\n",
        "\n",
        "pip install --upgrade matplotlib matplotlib-inline\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tfltpbZ_QLfZ"
      },
      "outputs": [],
      "source": [
        "#@title 2. Install pharokka and phold and phynteny\n",
        "\n",
        "#@markdown This cell installs pharokka and phold. It will take a few minutes. Please be patient\n",
        "\n",
        "PHAROKKA_VERSION=\"1.8.1\"\n",
        "PHOLD_VERSION=\"1.0.0\"\n",
        "PHYNTENY_VERSION=\"0.1.3\"\n",
        "\n",
        "# add paths\n",
        "import sys\n",
        "sys.path.append(\"/usr/local/bin\")\n",
        "\n",
        "# Update environment variables for shell usage\n",
        "import os\n",
        "os.environ[\"PATH\"] = \"/usr/local/bin:\" + os.environ[\"PATH\"]\n",
        "\n",
        "# create envs\n",
        "# pharokka isn't compatible with Python 3.13 (Google Colab default)\n",
        "# so it needs a separate env\n",
        "from pathlib import Path\n",
        "flag_file = Path(\"PHAROKKA_PHOLD_PHYNTENY_READY\")\n",
        "if not flag_file.exists():\n",
        "  !conda install -y -c bioconda phold=={PHOLD_VERSION} pharokka=={PHAROKKA_VERSION} pytorch=*=cuda* phynteny_transformer=={PHYNTENY_VERSION}\n",
        "  # Touch the flag file\n",
        "  flag_file.touch()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Sjdpu6R-Kig9"
      },
      "outputs": [],
      "source": [
        "#@title 3. Download pharokka phold databases\n",
        "\n",
        "#@markdown This cell downloads the pharokka then the phold database. It will take some time (10-15 minutes probably depending on Zenodo's traffic). Please be patient. Perhaps go for a walk or have a coffee or tea.\n",
        "\n",
        "\n",
        "%%time\n",
        "\n",
        "print(\"Downloading pharokka database. This will take a few minutes. Please be patient :)\")\n",
        "!install_databases.py -o pharokka_db\n",
        "\n",
        "\n",
        "print(\"Downloading phold database. This will take a few minutes. Please be patient :)\")\n",
        "!phold install -d phold_db -t 8 --foldseek_gpu\n",
        "\n",
        "\n",
        "print(\"Downloading phynteny database. This will take a few minutes. Please be patient :)\")\n",
        "!install_models -o  phynteny_models\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "9QfjP3q-Q04f",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title 4. Run Pharokka\n",
        "\n",
        "#@markdown First, upload your phage(s) as a nucleotide input FASTA file\n",
        "\n",
        "#@markdown Click on the folder icon to the left and use the file upload button.\n",
        "\n",
        "#@markdown Once it is uploaded, write the file name in the INPUT_FILE field on the right.\n",
        "\n",
        "#@markdown Then provide a directory for pharokka's output using PHAROKKA_OUT_DIR.\n",
        "#@markdown The default is 'output_pharokka'.\n",
        "\n",
        "#@markdown Then type in a gene prediction tool for pharokka.\n",
        "#@markdown Please choose either 'phanotate', 'prodigal', or 'prodigal-gv'.\n",
        "\n",
        "#@markdown You can also provide a prefix for your output files with PHAROKKA_PREFIX.\n",
        "#@markdown If you provide nothing it will default to 'pharokka'.\n",
        "\n",
        "#@markdown You can also provide a locus tag for your output files.\n",
        "#@markdown If you provide nothing it will generate a random locus tag.\n",
        "\n",
        "#@markdown You can click FAST to turn off --fast.\n",
        "#@markdown By default it is True so that Pharokka runs faster in the Colab environment.\n",
        "\n",
        "#@markdown You can click META to turn on --meta if you have multiple phages in your input.\n",
        "\n",
        "#@markdown You can click META_HMM to turn on --meta_hmm.\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory.\n",
        "#@markdown This may be useful if your earlier pharokka run has crashed for whatever reason.\n",
        "\n",
        "#@markdown The results of Pharokka will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is PHAROKKA_OUT_DIR.zip, where PHAROKKA_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "import sys\n",
        "import subprocess\n",
        "import zipfile\n",
        "INPUT_FILE = '' #@param {type:\"string\"}\n",
        "\n",
        "if os.path.exists(INPUT_FILE):\n",
        "    print(f\"Input file {INPUT_FILE} exists\")\n",
        "else:\n",
        "    print(f\"Error: File {INPUT_FILE} does not exist\")\n",
        "    print(f\"Please check the spelling and that you have uploaded it correctly\")\n",
        "    sys.exit(1)\n",
        "\n",
        "PHAROKKA_OUT_DIR = 'output_pharokka'  #@param {type:\"string\"}\n",
        "GENE_PREDICTOR = 'phanotate'  #@param {type:\"string\"}\n",
        "allowed_gene_predictors = ['phanotate', 'prodigal', 'prodigal-gv']\n",
        "# Check if the input parameter is valid\n",
        "if GENE_PREDICTOR.lower() not in allowed_gene_predictors:\n",
        "    raise ValueError(\"Invalid GENE_PREDICTOR. Please choose from: 'phanotate', 'prodigal', 'prodigal-gv'.\")\n",
        "\n",
        "PHAROKKA_PREFIX = 'pharokka'  #@param {type:\"string\"}\n",
        "LOCUS_TAG = 'Default'  #@param {type:\"string\"}\n",
        "FAST = True  #@param {type:\"boolean\"}\n",
        "META = False  #@param {type:\"boolean\"}\n",
        "META_HMM = False  #@param {type:\"boolean\"}\n",
        "FORCE = True  #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "# Construct the command\n",
        "# need to suppress PYTHONWARNINGS for phanotate version handling\n",
        "command = (\n",
        "    f'pharokka.py -d pharokka_db -i {INPUT_FILE} -t 4 '\n",
        "    f'-o {PHAROKKA_OUT_DIR} -p {PHAROKKA_PREFIX} -l {LOCUS_TAG} -g {GENE_PREDICTOR}'\n",
        ")\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "\n",
        "if FAST is True:\n",
        "  command = f\"{command} --fast\"\n",
        "\n",
        "if META is True:\n",
        "  command = f\"{command} -m\"\n",
        "\n",
        "if META_HMM is True:\n",
        "  command = f\"{command} --meta_hmm\"\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running pharokka\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"pharokka completed successfully.\")\n",
        "    print(f\"Your output is in {PHAROKKA_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{PHAROKKA_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(PHAROKKA_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), PHAROKKA_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5. Run phold\n",
        "\n",
        "#@markdown This cell will run phold on the output of cell 3's Pharokka run\n",
        "\n",
        "#@markdown You do not need to provide any further input files\n",
        "\n",
        "#@markdown You can now provide a directory for phold's output with PHOLD_OUT_DIR.\n",
        "#@markdown The default is 'output_phold'.\n",
        "\n",
        "#@markdown You can also provide a prefix for your output files with PHOLD_PREFIX.\n",
        "#@markdown If you provide nothing it will default to 'phold'.\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory with .\n",
        "#@markdown This may be useful if your earlier phold run has crashed for whatever reason.\n",
        "\n",
        "#@markdown If your input has multiple phages, you can click SEPARATE.\n",
        "#@markdown This will output separate GenBank files in the output directory.\n",
        "\n",
        "#@markdown The results of Phold will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is PHOLD_OUT_DIR.zip, where PHOLD_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "import subprocess\n",
        "import zipfile\n",
        "\n",
        "# phold input is pharokka output\n",
        "PHOLD_INPUT = f\"{PHAROKKA_OUT_DIR}/{PHAROKKA_PREFIX}.gbk\"\n",
        "PHOLD_OUT_DIR = 'output_phold'  #@param {type:\"string\"}\n",
        "PHOLD_PREFIX = 'phold'  #@param {type:\"string\"}\n",
        "FORCE = True  #@param {type:\"boolean\"}\n",
        "SEPARATE = False  #@param {type:\"boolean\"}\n",
        "\n",
        "# Construct the command\n",
        "command = f\"phold run -i {PHOLD_INPUT} -t 4 -o {PHOLD_OUT_DIR} -p {PHOLD_PREFIX} -d phold_db --foldseek_gpu\"\n",
        "\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "if SEPARATE is True:\n",
        "  command = f\"{command} --separate\"\n",
        "\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running phold\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"phold completed successfully.\")\n",
        "    print(f\"Your output is in {PHOLD_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{PHOLD_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(PHOLD_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), PHOLD_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "1eUYA415PNRX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BlPLP3vlhWYs"
      },
      "outputs": [],
      "source": [
        "#@title 6. Run Phynteny\n",
        "\n",
        "#@markdown This cell will run phynteny on the output of cell 5's Phold run to predict the function of remaining hypothetical proteins\n",
        "\n",
        "#@markdown You do not need to provide any further input files\n",
        "\n",
        "#@markdown You can now provide a directory for phynteny's output with PHYNTENY_OUT_DIR.\n",
        "#@markdown The default is 'output_phynteny'.\n",
        "\n",
        "#@markdown You can click FORCE to overwrite the output directory with .\n",
        "#@markdown This may be useful if your phynteny run has crashed for whatever reason.\n",
        "\n",
        "#@markdown The results of Phynteny will be in the folder icon on the left hand panel.\n",
        "#@markdown Additionally, it will be zipped so you can download the whole directory.\n",
        "\n",
        "#@markdown The file to download is PHYNTENY_OUT_DIR.zip, where PHYNTENY_OUT_DIR is what you provided\n",
        "\n",
        "#@markdown If you do not see the output directory,\n",
        "#@markdown refresh the window by either clicking the folder with the refresh icon below \"Files\"\n",
        "#@markdown or double click and select \"Refresh\".\n",
        "\n",
        "\n",
        "%%time\n",
        "import os\n",
        "import subprocess\n",
        "import zipfile\n",
        "\n",
        "# phynteny input is pharokka output\n",
        "PHYNTENY_INPUT = f\"{PHOLD_OUT_DIR}/{PHOLD_PREFIX}.gbk\"\n",
        "PHYNTENY_OUT_DIR = 'output_phynteny'  #@param {type:\"string\"}\n",
        "FORCE = False  #@param {type:\"boolean\"}\n",
        "\n",
        "# Construct the command\n",
        "command = f\"phynteny_transformer -m  /content/phynteny_models/models -o {PHYNTENY_OUT_DIR} {PHYNTENY_INPUT}\"\n",
        "if FORCE is True:\n",
        "  command = f\"{command} -f\"\n",
        "\n",
        "\n",
        "# Execute the command\n",
        "try:\n",
        "    print(\"Running phynteny\")\n",
        "    subprocess.run(command, shell=True, check=True)\n",
        "    print(\"phynteny completed successfully.\")\n",
        "    print(f\"Your output is in {PHYNTENY_OUT_DIR}.\")\n",
        "    print(f\"Zipping the output directory so you can download it all in one go.\")\n",
        "\n",
        "    zip_filename = f\"{PHYNTENY_OUT_DIR}.zip\"\n",
        "\n",
        "    # Zip the contents of the output directory\n",
        "    with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        for root, dirs, files in os.walk(PHYNTENY_OUT_DIR):\n",
        "            for file in files:\n",
        "                zipf.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), PHYNTENY_OUT_DIR))\n",
        "    print(f\"Output directory has been zipped to {zip_filename}\")\n",
        "\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(f\"Error occurred: {e}\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "authorship_tag": "ABX9TyPUxBtfVRwU7FqMXK2cz1B1",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "nbformat": 4,
  "nbformat_minor": 0
}